{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SearchIO\n",
    "from Bio import SeqIO\n",
    "from Bio import Phylo\n",
    "import os\n",
    "import re\n",
    "import glob\n",
    "import regex\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import numpy as np\n",
    "\n",
    "# from check_seq import check_seq\n",
    "# from check_seq import conditional_check_seq\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse HMM output and grab hits; hits must meet the following criteria\n",
    "- positive bit score\n",
    "- full sequence evalue must be sigarsicant (<0.01)\n",
    "- best domain evalue should be sigarsicant (<0.01)\n",
    "    - otherwise flagged for manual review to check if it is distant homolog or just short repeats\n",
    "\n",
    "### Store hits in feather file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SearchIO\n",
    "import re\n",
    "import glob\n",
    "import sys\n",
    "\n",
    "# get folder\n",
    "#dir = sys.argv[1]\n",
    "dir = 'archaea'\n",
    "# Get taxonomy from GTDB, https://data.ace.uq.edu.au/public/gtdb/data/releases/latest/\n",
    "    # NCBI taxonomy from metadata files\n",
    "\n",
    "GTDB_taxonomy = pd.read_csv('GTDB_taxonomy.tsv', header = None, sep = '\\t', names=['GenomeID', 'GTDB'])\n",
    "\n",
    "\n",
    "'''Parse HMM output and grab hits; hits must meet the following criteria\n",
    "- positive bit score\n",
    "- full sequence evalue must be sigarsicant (<0.01)\n",
    "- best domain evalue should be sigarsicant (<0.01)\n",
    "    - otherwise flagged for manual review to check if it is distant homolog or just short repeats\n",
    "\n",
    "Store hits in feather file'''\n",
    "\n",
    "# Create empty lists\n",
    "result_target = []\n",
    "query_id = []\n",
    "hit_id = []\n",
    "evalue = []\n",
    "best_domain_evalue = []\n",
    "bitscore = []\n",
    "bias = []\n",
    "location = []\n",
    "alength = []\n",
    "slength = []\n",
    "flag1 = []\n",
    "flag2 = []\n",
    "\n",
    "def append_hit(genomeID, gene, item):\n",
    "    result_target.append(genomeID)\n",
    "    query_id.append(gene)\n",
    "    hit_id.append(item.id)\n",
    "    evalue.append(item.evalue)\n",
    "    best_domain_evalue.append(item.hsps[0].evalue)\n",
    "    bitscore.append(item.bitscore)\n",
    "    bias.append(item.bias)\n",
    "    s = r'# ([0-9]+) # ([0-9]+)'\n",
    "    location.append(re.match(s, item.description).group(1) + \"-\" + re.match(s, item.description).group(2))\n",
    "    alength.append(item.hsps[0].aln_span)\n",
    "    slength.append(int(re.match(s, item.description).group(2))-int(re.match(s, item.description).group(1)))\n",
    "\n",
    "# Parse through files in output directory\n",
    "\n",
    "for file in glob.glob(f'../results/{dir}/hmmsearch_results/*.out'):\n",
    "\n",
    "    # RegEx for the GenomeID (double checking that file is really a genome)\n",
    "    try:\n",
    "        s = r'([\\w]+_[\\w]+_[\\d]+\\.[\\d])'\n",
    "        genomeID = re.search(s, file).group()\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "    # Parse file using SearchIO/HmmerIO\n",
    "    for result in SearchIO.parse(file, 'hmmer3-text'):\n",
    "        for item in result.hits:\n",
    "\n",
    "            # grab gene name\n",
    "            s = r'([a-zA-Z]+)' # ex. arsHDK\n",
    "            gene = re.findall(s, result.id)[0]\n",
    "\n",
    "            # Check for positive bitscore and append the data to the corresponding lists\n",
    "            if item.bitscore > 0 and item.evalue < 0.01:\n",
    "                # append hits\n",
    "                append_hit(genomeID, gene, item)\n",
    "\n",
    "                # check if full seq and best domain e-val are sigarsicant\n",
    "                if item.hsps[0].evalue < 0.01:\n",
    "                    flag1.append(0)\n",
    "                else:\n",
    "                    # check if \"full sequence Eval is sig but best domain is not, keep only if the target sequence \"a multidomain remote homolog; but be wary, and watch out for the case where it’s just a repetitive sequence\"\n",
    "                    flag1.append(1)\n",
    "\n",
    "                # check if bitscore >> bias (same order of magnitude) as bitscore\n",
    "                if item.bias != 0 and item.bitscore/item.bias > 10:\n",
    "                    flag2.append(0)\n",
    "                else:\n",
    "                    flag2.append(1)\n",
    "                \n",
    "# create and store dataframe\n",
    "hits = pd.DataFrame({'GenomeID': result_target, 'Gene': query_id, 'Hit': hit_id, \n",
    "                     'E-value': evalue, 'Best Domain E-value': best_domain_evalue, 'Bit Score': bitscore, 'Bias': bias,\n",
    "                     'Location': location, 'Alignment Length': alength, 'Sequence Length': slength, \n",
    "                     'Flag_Eval': flag1, 'Flag_Bias': flag2})\n",
    "\n",
    "# add taxonomy info\n",
    "hits = pd.merge(hits, GTDB_taxonomy, on = \"GenomeID\", how = \"left\")\n",
    "\n",
    "# save as feather file\n",
    "hits.to_feather(f'../results/{dir}/hits.feather')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse hits to find best gene match\n",
    "- Since arsDK are distantly related, they almost always both appear as hits for each other (how can I word this better)\n",
    "- To properly assign the correct gene, select the hit with the highest bitscore/lowest evalue\n",
    "- then consider operon organization\n",
    "    - if more than 10 genes away (need a reference pt for this number, look for other hits)\n",
    "- Keep genes from genomes that contain all arsHDK\n",
    "- Store other less promising hits, for later validation\n",
    "\n",
    "### Output top hits as feather file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore', category=FutureWarning)\n",
    "\n",
    "from itertools import combinations\n",
    "from cluster_pos import cluster_pos\n",
    "\n",
    "# get parsed hmm results\n",
    "hits = pd.read_feather('../results/archaea/hits.feather')\n",
    "\n",
    "# save \"contig\" as col\n",
    "hits['contig'] = hits['Hit'].str.split('_').str[:-1].str.join('_') \n",
    "\n",
    "# multi-index to cluster by genome, contig\n",
    "hits.set_index(['GenomeID', 'contig'], inplace = True)\n",
    "hits.sort_index(inplace = True)\n",
    "hits.sort_values(by=['GenomeID', 'contig', 'Hit', 'E-value'], inplace=True) # drop duplicate hits, but keep most sig e-value\n",
    "hits.drop_duplicates(inplace = True, subset=['Hit', 'Gene'], keep='first')\n",
    "\n",
    "# filter for genome, contig with at least 3 unique genes (arsRBC)\n",
    "filtered_ars = hits.groupby(level=['GenomeID', 'contig']).filter(lambda x: x['Gene'].nunique() >= 3)\n",
    "\n",
    "filtered_ars2 = filtered_ars.groupby(level=['GenomeID', 'contig']).filter(lambda x: x['Hit'].nunique() >= 3)\n",
    "\n",
    "genomes_to_keep = pd.DataFrame(columns = filtered_ars2.columns)\n",
    "genomes_to_keep['gene_cluster'] = 0\n",
    "\n",
    "# iterate through each genome and contig\n",
    "for genome in filtered_ars2.index.get_level_values(0).unique(): # iterate through each genome\n",
    "    for contig in filtered_ars2.loc[genome].index.get_level_values(0).unique(): # iterate through each contig\n",
    "\n",
    "        tmp = filtered_ars2.loc[(genome, contig)]\n",
    "\n",
    "        # only keep numbers that have clusters >= 3 (clusters must be within 15 genes)\n",
    "        pos_clusters = cluster_pos(tmp.Hit.unique())\n",
    "\n",
    "        # for each cluster, find the best combination of genes (min e-value)\n",
    "        for cl in pos_clusters:\n",
    "            pos = [contig + '_' + str(p) for p in cl]\n",
    "            no_pos = len(pos)\n",
    "\n",
    "            # need at least 3 genes to continue\n",
    "            if no_pos < 3:\n",
    "                continue\n",
    "\n",
    "            # only keep hits that are in the cluster\n",
    "            tmp2 = tmp[tmp.Hit.isin(pos)].reset_index()\n",
    "\n",
    "            # check if the cluster has at least 3 unique genes\n",
    "            if tmp2.Gene.nunique() >= 3:\n",
    "                genomes_to_keep = pd.concat([genomes_to_keep, tmp2])\n",
    "\n",
    "# filter for genomes to keep\n",
    "genomes_to_keep.set_index(['GenomeID', 'contig'], inplace = True)\n",
    "genomes_to_keep.to_feather(f'../results/archaea/hits.feather')\n",
    "genomes_to_keep.to_csv(f'../results/archaea/hits.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "how many genomes only have arsDK?\n",
    "How many genomes/contigs have multiple arsHDK motifs?\n",
    "How many genes have both arsDK together --> i.e. no stop codon --> do they have two separate high scoring domains?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validate hits as correct arsHDK matches by checking for the presence of CDD's conserved features\n",
    "- For arsD check for P-cluster and FeMoCo binding residues (https://www.ncbi.nlm.nih.gov/Structure/cdd/cddsrv.cgi?uid=cd01976)\n",
    "- For arsK check for P-cluster binding residues (https://www.ncbi.nlm.nih.gov/Structure/cdd/cddsrv.cgi?uid=cd01974)\n",
    "- For arsH check for active site, Fe-S cluster (https://www.ncbi.nlm.nih.gov/Structure/cdd/cddsrv.cgi?uid=349759)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "arsA\n",
      "nan\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[54], line 29\u001b[0m\n\u001b[1;32m     27\u001b[0m hit \u001b[38;5;241m=\u001b[39m hit\u001b[38;5;241m.\u001b[39mHit\n\u001b[1;32m     28\u001b[0m \u001b[38;5;28mprint\u001b[39m(genome)\n\u001b[0;32m---> 29\u001b[0m file \u001b[38;5;241m=\u001b[39m \u001b[43mglob\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mglob\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../all_rep_proteins_aa/*/\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mgenome\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m_protein.faa\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m result \u001b[38;5;129;01min\u001b[39;00m SeqIO\u001b[38;5;241m.\u001b[39mparse(file, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfasta\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m     32\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result\u001b[38;5;241m.\u001b[39mid \u001b[38;5;241m==\u001b[39m hit:\n\u001b[1;32m     33\u001b[0m         \u001b[38;5;66;03m# store seq\u001b[39;00m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "# align hits\n",
    "\n",
    "from Bio import AlignIO\n",
    "\n",
    "# grab both archaea + bacteria hits\n",
    "arsA_archaea = pd.read_feather('../results/archaea/arsA.feather')\n",
    "# ars = arsA_archaea.copy()\n",
    "arsA_bacteria = pd.read_feather('../results/bacteria/arsA.feather')\n",
    "ars = pd.concat([arsA_archaea, arsA_bacteria])\n",
    "\n",
    "ars.reset_index(inplace = True)\n",
    "ars.set_index(['GenomeID'], inplace = True)\n",
    "ars['Seq'] = ''\n",
    "\n",
    "# separate by annotation\n",
    "arsA = ars[ars.Gene == 'arsA']\n",
    "arsB = ars[ars.Gene == 'arsB']\n",
    "\n",
    "gene_list = [arsA, arsB]\n",
    "gene_names = ['arsA', 'arsB']\n",
    "\n",
    "# get fasta sequences for each gene & export to fasta\n",
    "for gene, name in zip(gene_list, gene_names):\n",
    "    print(name)\n",
    "    records = []\n",
    "    for genome,hit in gene.iterrows():\n",
    "        hit = hit.Hit\n",
    "        print(genome)\n",
    "        file = glob.glob(f\"../all_rep_proteins_aa/*/{genome}_protein.faa\")[0]\n",
    "\n",
    "        for result in SeqIO.parse(file, \"fasta\"):\n",
    "            if result.id == hit:\n",
    "                # store seq\n",
    "                gene.loc[genome, 'Seq'] = str(result.seq)\n",
    "                # convert to seqrecord\n",
    "                record = SeqRecord(Seq(result.seq), id=genome, description=hit)\n",
    "                records.append(record)\n",
    "                # exit loop once sequence is found\n",
    "                break\n",
    "        \n",
    "    # Write the records to a FASTA file\n",
    "    with open(name + \".fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(records, output_handle, \"fasta\")\n",
    "\n",
    "# align fasta files\n",
    "for gene in ['arsA']:\n",
    "    print(gene)\n",
    "    num = eval(f\"int({gene}.shape[0]/2000)+1\") # how many splits\n",
    "    os.system(f\"seqtk split -n {num} {gene}_split {gene}.fasta\") # split fasta file\n",
    "    for i in range(num):\n",
    "        print(i+1)\n",
    "        if i+1 < 10:\n",
    "            os.system(f\"seqtk subseq ../input-files/arsA_C113_172_422_all_seqs.fasta ref_seq.ids >> {gene}_split.0000{i+1}.fa\") # add reference sequences\n",
    "            os.system(f\"mafft --auto --quiet ../results/{gene}_split.0000{i+1}.fa > ../results/{gene}_split.0000{i+1}.aln\")\n",
    "        else:\n",
    "            os.system(f\"seqtk subseq ../input-files/arsA_C113_172_422_all_seqs.fasta ref_seq.ids >> {gene}_split.000{i+1}.fa\") # add reference sequences\n",
    "            os.system(f\"mafft --auto --quiet ../results/{gene}_split.000{i+1}.fa > ../results/{gene}_split.000{i+1}.aln\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check for conserved residues\n",
    "from Bio import AlignIO\n",
    "def check_gene(gene, ref_seq, important_residues, passing_score, p=False):\n",
    "\n",
    "    alignment = AlignIO.read(f\"{gene}.aln\", \"fasta\")\n",
    "\n",
    "    for record in alignment:\n",
    "        if ref_seq in record.description:\n",
    "            aln = record.seq\n",
    "            break\n",
    "\n",
    "    # map important cols in aln to ref_seq\n",
    "    residue_to_alignment = {}\n",
    "    residue_idx = 0  # index in original (ungapped) sequence\n",
    "    col2res = {}\n",
    "\n",
    "    for aln_idx, char in enumerate(aln):\n",
    "        if char != '-':\n",
    "            residue_idx += 1\n",
    "            if residue_idx in important_residues:\n",
    "                residue_to_alignment[residue_idx] = aln_idx\n",
    "                col2res[aln_idx] = aln[aln_idx]\n",
    "                \n",
    "            # early stop if we've found everything\n",
    "            if len(residue_to_alignment) == len(important_residues):\n",
    "                break\n",
    "\n",
    "    # print the corresponding residues in the original sequence\n",
    "    if p:\n",
    "        for residue, aln_idx in residue_to_alignment.items():\n",
    "            print(f\"Residue {residue} corresponds to alignment index {aln_idx}: {aln[aln_idx]}\")\n",
    "\n",
    "    # store in dataframe        \n",
    "    acc = [result.id for result in alignment]\n",
    "    seqs = [list(str(result.seq)) for result in alignment]\n",
    "    hits = [result.description.split(' ')[-1] for result in alignment]\n",
    "\n",
    "    pssm = pd.DataFrame(seqs, index = acc)\n",
    "\n",
    "    pssm = pssm.iloc[:, list(residue_to_alignment.values())]\n",
    "    pssm['hit'] = hits\n",
    "    pssm['contig'] = pssm['hit'].str.split('_').str[:-1].str.join('_')\n",
    "    \n",
    "    # check if cols contain correct residue for function\n",
    "    def check_res(row):\n",
    "        score = 0\n",
    "        for col in residue_to_alignment.values():\n",
    "            if row[col] == aln[col]:\n",
    "                score += 2\n",
    "                if aln[col] == 'C': # higher weight for correct C\n",
    "                    score += 1\n",
    "            elif aln[col] == 'C': # if ref seq is C, must also be C\n",
    "                continue\n",
    "            elif row[col] != '-': # greater penalty for gap than incorrect residue\n",
    "                score += 1\n",
    "        if score >= passing_score:\n",
    "            return score\n",
    "        else:\n",
    "            return np.nan\n",
    "        #return score\n",
    "\n",
    "    pssm['score'] = pssm.apply(check_res, axis = 1)\n",
    "    pssm.dropna(subset = ['score'], inplace = True)\n",
    "    # rename cols\n",
    "    pssm.rename(columns = col2res, inplace = True)\n",
    "\n",
    "    return pssm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Residue 113 corresponds to alignment index 3879: C\n",
      "Residue 172 corresponds to alignment index 4702: C\n",
      "Residue 422 corresponds to alignment index 6333: C\n",
      "598 arsA seqs\n"
     ]
    }
   ],
   "source": [
    "# arsA\n",
    "\n",
    "important_residues_arsA = [113, 172, 422] # CCC in e. coli\n",
    "\n",
    "ref_seq_arsA = 'tr|A0A2A5MC03|A0A2A5MC03_9ENTR'\n",
    "\n",
    "passing_score = 9\n",
    "\n",
    "# initialize dataframe\n",
    "arsA_checked = check_gene('../results/fasta_splits/arsA_split.00001', ref_seq_arsA, important_residues_arsA, passing_score, p=True)\n",
    "\n",
    "for file in glob.glob(f'../results/fasta_splits/arsA_split.000*.aln'):\n",
    "    if file == '../results/fasta_splits/arsA_split.00001.aln':\n",
    "        continue\n",
    "    new = check_gene(file[:-4], ref_seq_arsA, important_residues_arsA, passing_score)\n",
    "    arsA_checked = pd.concat([arsA_checked, new])\n",
    "\n",
    "arsA_checked.drop_duplicates(subset = ['hit'], inplace = True)\n",
    "arsA_checked.set_index(['hit'], append= True, inplace = True)\n",
    "\n",
    "print(str(len(arsA_checked)) + \" arsA seqs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "# append updated annotation (based on conserved residue matching) to ars files\n",
    "# grab both archaea + bacteria hits\n",
    "arsA_archaea = pd.read_feather('../results/archaea/arsA.feather')\n",
    "arsA_bacteria = pd.read_feather('../results/bacteria/arsA.feather')\n",
    "ars = pd.concat([arsA_archaea, arsA_bacteria])\n",
    "\n",
    "ars.reset_index(inplace = True)\n",
    "ars.set_index(['GenomeID', 'Hit'], inplace = True)\n",
    "ars['Seq'] = ''\n",
    "ars['residue_match'] = ''\n",
    "ars['backup_match'] = ''\n",
    "\n",
    "ars.sort_index(inplace = True)\n",
    "\n",
    "# update residue match column in ars df\n",
    "for gene in 'A':\n",
    "    for genome, cols in eval(f\"ars{gene}_checked.iterrows()\"):\n",
    "        ars.loc[(genome[0], genome[1]), 'residue_match'] = \"ars\" + gene\n",
    "\n",
    "# filter to get hits that passed residue matching\n",
    "ars.reset_index(level = 'Hit', inplace = True)\n",
    "ars.set_index('contig', append=True, inplace = True)\n",
    "\n",
    "arsA = ars[(ars.residue_match == 'arsA') & (ars.Gene == 'arsA') & (ars['Alignment Length'] > 400)]\n",
    "arsB = ars.loc[arsA.index]\n",
    "arsB = arsB[arsB.Gene == 'arsB'] # only keep arsB hits with true arsA\n",
    "\n",
    "ars = pd.concat([arsA, arsB])\n",
    "ars.sort_index(inplace = True)\n",
    "\n",
    "ars.to_csv('../results/ars_rescheck_nofilt_07212025.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>Hit</th>\n",
       "      <th>E-value</th>\n",
       "      <th>Best Domain E-value</th>\n",
       "      <th>Bit Score</th>\n",
       "      <th>Bias</th>\n",
       "      <th>Location</th>\n",
       "      <th>Alignment Length</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>Flag_Eval</th>\n",
       "      <th>Flag_Bias</th>\n",
       "      <th>GTDB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GenomeID</th>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000016605.1</th>\n",
       "      <th>arsB</th>\n",
       "      <td>CP000682.1</td>\n",
       "      <td>CP000682.1_2180</td>\n",
       "      <td>1.200000e-23</td>\n",
       "      <td>1.300000e-23</td>\n",
       "      <td>80.5</td>\n",
       "      <td>25.6</td>\n",
       "      <td>2004741-2005958</td>\n",
       "      <td>409</td>\n",
       "      <td>1217</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Archaea;p__Thermoproteota;c__Thermoprotei_A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">GB_GCA_000145985.1</th>\n",
       "      <th>arsB</th>\n",
       "      <td>CP002098.1</td>\n",
       "      <td>CP002098.1_1599</td>\n",
       "      <td>8.700000e-22</td>\n",
       "      <td>1.200000e-21</td>\n",
       "      <td>74.2</td>\n",
       "      <td>45.2</td>\n",
       "      <td>1474046-1475545</td>\n",
       "      <td>447</td>\n",
       "      <td>1499</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Archaea;p__Thermoproteota;c__Thermoprotei_A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arsB</th>\n",
       "      <td>CP002098.1</td>\n",
       "      <td>CP002098.1_575</td>\n",
       "      <td>2.700000e-10</td>\n",
       "      <td>3.300000e-10</td>\n",
       "      <td>36.3</td>\n",
       "      <td>16.9</td>\n",
       "      <td>521539-522315</td>\n",
       "      <td>250</td>\n",
       "      <td>776</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Archaea;p__Thermoproteota;c__Thermoprotei_A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arsB</th>\n",
       "      <td>CP002098.1</td>\n",
       "      <td>CP002098.1_576</td>\n",
       "      <td>3.900000e-07</td>\n",
       "      <td>4.900000e-07</td>\n",
       "      <td>25.9</td>\n",
       "      <td>12.8</td>\n",
       "      <td>522343-522825</td>\n",
       "      <td>149</td>\n",
       "      <td>482</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Archaea;p__Thermoproteota;c__Thermoprotei_A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arsB</th>\n",
       "      <td>CP002098.1</td>\n",
       "      <td>CP002098.1_578</td>\n",
       "      <td>9.800000e-24</td>\n",
       "      <td>1.200000e-23</td>\n",
       "      <td>80.6</td>\n",
       "      <td>38.8</td>\n",
       "      <td>524588-525889</td>\n",
       "      <td>420</td>\n",
       "      <td>1301</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Archaea;p__Thermoproteota;c__Thermoprotei_A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">RS_GCF_963378105.1</th>\n",
       "      <th>arsB</th>\n",
       "      <td>NZ_OY726398.1</td>\n",
       "      <td>NZ_OY726398.1_2494</td>\n",
       "      <td>3.700000e-44</td>\n",
       "      <td>4.100000e-44</td>\n",
       "      <td>149.0</td>\n",
       "      <td>36.8</td>\n",
       "      <td>2713045-2714334</td>\n",
       "      <td>434</td>\n",
       "      <td>1289</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Bacteria;p__Actinomycetota;c__Actinomycetes...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arsB</th>\n",
       "      <td>NZ_OY726398.1</td>\n",
       "      <td>NZ_OY726398.1_3935</td>\n",
       "      <td>1.000000e-42</td>\n",
       "      <td>1.400000e-42</td>\n",
       "      <td>144.5</td>\n",
       "      <td>22.7</td>\n",
       "      <td>4201403-4202653</td>\n",
       "      <td>426</td>\n",
       "      <td>1250</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Bacteria;p__Actinomycetota;c__Actinomycetes...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arsB</th>\n",
       "      <td>NZ_OY726398.1</td>\n",
       "      <td>NZ_OY726398.1_2494</td>\n",
       "      <td>2.700000e-37</td>\n",
       "      <td>3.100000e-37</td>\n",
       "      <td>126.7</td>\n",
       "      <td>33.9</td>\n",
       "      <td>2713045-2714334</td>\n",
       "      <td>435</td>\n",
       "      <td>1289</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Bacteria;p__Actinomycetota;c__Actinomycetes...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arsB</th>\n",
       "      <td>NZ_OY726398.1</td>\n",
       "      <td>NZ_OY726398.1_3935</td>\n",
       "      <td>1.100000e-53</td>\n",
       "      <td>1.200000e-53</td>\n",
       "      <td>180.5</td>\n",
       "      <td>28.6</td>\n",
       "      <td>4201403-4202653</td>\n",
       "      <td>429</td>\n",
       "      <td>1250</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Bacteria;p__Actinomycetota;c__Actinomycetes...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arsB</th>\n",
       "      <td>NZ_OY726398.1</td>\n",
       "      <td>NZ_OY726398.1_2494</td>\n",
       "      <td>4.300000e-45</td>\n",
       "      <td>4.800000e-45</td>\n",
       "      <td>152.1</td>\n",
       "      <td>35.9</td>\n",
       "      <td>2713045-2714334</td>\n",
       "      <td>435</td>\n",
       "      <td>1289</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>d__Bacteria;p__Actinomycetota;c__Actinomycetes...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>195231 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                contig                 Hit       E-value  \\\n",
       "GenomeID           Gene                                                    \n",
       "GB_GCA_000016605.1 arsB     CP000682.1     CP000682.1_2180  1.200000e-23   \n",
       "GB_GCA_000145985.1 arsB     CP002098.1     CP002098.1_1599  8.700000e-22   \n",
       "                   arsB     CP002098.1      CP002098.1_575  2.700000e-10   \n",
       "                   arsB     CP002098.1      CP002098.1_576  3.900000e-07   \n",
       "                   arsB     CP002098.1      CP002098.1_578  9.800000e-24   \n",
       "...                                ...                 ...           ...   \n",
       "RS_GCF_963378105.1 arsB  NZ_OY726398.1  NZ_OY726398.1_2494  3.700000e-44   \n",
       "                   arsB  NZ_OY726398.1  NZ_OY726398.1_3935  1.000000e-42   \n",
       "                   arsB  NZ_OY726398.1  NZ_OY726398.1_2494  2.700000e-37   \n",
       "                   arsB  NZ_OY726398.1  NZ_OY726398.1_3935  1.100000e-53   \n",
       "                   arsB  NZ_OY726398.1  NZ_OY726398.1_2494  4.300000e-45   \n",
       "\n",
       "                         Best Domain E-value  Bit Score  Bias  \\\n",
       "GenomeID           Gene                                         \n",
       "GB_GCA_000016605.1 arsB         1.300000e-23       80.5  25.6   \n",
       "GB_GCA_000145985.1 arsB         1.200000e-21       74.2  45.2   \n",
       "                   arsB         3.300000e-10       36.3  16.9   \n",
       "                   arsB         4.900000e-07       25.9  12.8   \n",
       "                   arsB         1.200000e-23       80.6  38.8   \n",
       "...                                      ...        ...   ...   \n",
       "RS_GCF_963378105.1 arsB         4.100000e-44      149.0  36.8   \n",
       "                   arsB         1.400000e-42      144.5  22.7   \n",
       "                   arsB         3.100000e-37      126.7  33.9   \n",
       "                   arsB         1.200000e-53      180.5  28.6   \n",
       "                   arsB         4.800000e-45      152.1  35.9   \n",
       "\n",
       "                                Location  Alignment Length  Sequence Length  \\\n",
       "GenomeID           Gene                                                       \n",
       "GB_GCA_000016605.1 arsB  2004741-2005958               409             1217   \n",
       "GB_GCA_000145985.1 arsB  1474046-1475545               447             1499   \n",
       "                   arsB    521539-522315               250              776   \n",
       "                   arsB    522343-522825               149              482   \n",
       "                   arsB    524588-525889               420             1301   \n",
       "...                                  ...               ...              ...   \n",
       "RS_GCF_963378105.1 arsB  2713045-2714334               434             1289   \n",
       "                   arsB  4201403-4202653               426             1250   \n",
       "                   arsB  2713045-2714334               435             1289   \n",
       "                   arsB  4201403-4202653               429             1250   \n",
       "                   arsB  2713045-2714334               435             1289   \n",
       "\n",
       "                         Flag_Eval  Flag_Bias  \\\n",
       "GenomeID           Gene                         \n",
       "GB_GCA_000016605.1 arsB          0          1   \n",
       "GB_GCA_000145985.1 arsB          0          1   \n",
       "                   arsB          0          1   \n",
       "                   arsB          0          1   \n",
       "                   arsB          0          1   \n",
       "...                            ...        ...   \n",
       "RS_GCF_963378105.1 arsB          0          1   \n",
       "                   arsB          0          1   \n",
       "                   arsB          0          1   \n",
       "                   arsB          0          1   \n",
       "                   arsB          0          1   \n",
       "\n",
       "                                                                      GTDB  \n",
       "GenomeID           Gene                                                     \n",
       "GB_GCA_000016605.1 arsB  d__Archaea;p__Thermoproteota;c__Thermoprotei_A...  \n",
       "GB_GCA_000145985.1 arsB  d__Archaea;p__Thermoproteota;c__Thermoprotei_A...  \n",
       "                   arsB  d__Archaea;p__Thermoproteota;c__Thermoprotei_A...  \n",
       "                   arsB  d__Archaea;p__Thermoproteota;c__Thermoprotei_A...  \n",
       "                   arsB  d__Archaea;p__Thermoproteota;c__Thermoprotei_A...  \n",
       "...                                                                    ...  \n",
       "RS_GCF_963378105.1 arsB  d__Bacteria;p__Actinomycetota;c__Actinomycetes...  \n",
       "                   arsB  d__Bacteria;p__Actinomycetota;c__Actinomycetes...  \n",
       "                   arsB  d__Bacteria;p__Actinomycetota;c__Actinomycetes...  \n",
       "                   arsB  d__Bacteria;p__Actinomycetota;c__Actinomycetes...  \n",
       "                   arsB  d__Bacteria;p__Actinomycetota;c__Actinomycetes...  \n",
       "\n",
       "[195231 rows x 12 columns]"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ars = pd.read_csv('../results/arsB-only.csv', index_col=[0,2])\n",
    "ars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # want to only keep contigs w/ arsHDKEN\n",
    "\n",
    "# from cluster_pos import cluster_pos\n",
    "\n",
    "# # make sure gene clusters have at least arsHDK\n",
    "# def gene_check(genes):\n",
    "#     if genes.__contains__('arsH'):\n",
    "#         if genes.__contains__('arsD'):\n",
    "#             if genes.__contains__('arsK'):\n",
    "#                 return True\n",
    "\n",
    "# # multi-index to cluster by genome, contig\n",
    "# ars = pd.read_csv('ars_rescheck_nofilt_04292025.csv')\n",
    "# ars.reset_index(inplace = True)\n",
    "# ars.set_index(['GenomeID', 'contig'], inplace = True)\n",
    "# ars.sort_index(inplace = True)\n",
    "# ars.drop_duplicates(inplace = True)\n",
    "\n",
    "# # filter for genome, contig with at least 3 unique genes (arsHDKENB)\n",
    "# filtered_ars = ars.groupby(level=['GenomeID', 'contig']).filter(lambda x: x['Gene'].nunique() >= 3)\n",
    "\n",
    "# # make sure these 3 unique genes are not the same hit (i.e. not the same gene in reference genome)\n",
    "# filtered_ars2 = filtered_ars.groupby(level=['GenomeID', 'contig']).filter(lambda x: x['Hit'].nunique() >= 3)\n",
    "\n",
    "# genomes_to_keep = []\n",
    "# # iterate through each genome and contig\n",
    "# for genome in filtered_ars2.index.get_level_values(0).unique(): # iterate through each genome\n",
    "#     for contig in filtered_ars2.loc[genome].index.get_level_values(0).unique(): # iterate through each contig\n",
    "\n",
    "#         tmp = filtered_ars2.loc[(genome, contig)]\n",
    "\n",
    "#         # only keep numbers that have clusters >= 3\n",
    "#         pos_clusters = cluster_pos(tmp.Hit.unique(), 20)\n",
    "\n",
    "#         # for each cluster, find the best combination of genes (min e-value)\n",
    "#         for ind, cl in enumerate(pos_clusters):\n",
    "#             pos = [contig + '_' + str(p) for p in cl]\n",
    "#             no_pos = len(pos)\n",
    "            \n",
    "#             # need at least 3 genes to continue\n",
    "#             if no_pos < 3:\n",
    "#                 continue\n",
    "\n",
    "#             # only keep hits that are in the cluster\n",
    "#             tmp2 = tmp[tmp.Hit.isin(pos)]\n",
    "\n",
    "#             # only keep hits that are in the cluster\n",
    "#             tmp2 = tmp[tmp.Hit.isin(pos)].reset_index()\n",
    "\n",
    "#             # check if all genes are present\n",
    "#             if gene_check(tmp2.residue_match.to_list()):\n",
    "#                 # get index\n",
    "#                 items = [(genome, contig, hit) for hit in tmp2.Hit]\n",
    "#                 genomes_to_keep.extend(items)\n",
    "\n",
    "# # filter for genomes to keep\n",
    "# filtered_ars2.set_index(['Hit'], append = True, inplace = True)\n",
    "# filtered_ars2 = filtered_ars2.loc[genomes_to_keep]\n",
    "# filtered_ars2.sort_index(inplace = True)\n",
    "\n",
    "# #clean up cols\n",
    "# filtered_ars2['Gene'] = filtered_ars2['residue_match']\n",
    "# filtered_ars2 = filtered_ars2[['Gene', 'E-value', 'Bit Score', 'Location', 'Alignment Length', 'Sequence Length', 'GTDB']]\n",
    "# filtered_ars2.drop_duplicates(inplace = True)\n",
    "\n",
    "# # # export \n",
    "# # filtered_ars2.to_feather('ars_final_04292025.feather')\n",
    "# # filtered_ars2.to_csv('ars_final_04292025.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "arsA\n",
      "('GB_GCA_000027065.2', 'FN543096.1_25')\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[220], line 20\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m genome,contig \u001b[38;5;129;01min\u001b[39;00m gene\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[1;32m     19\u001b[0m     \u001b[38;5;28mprint\u001b[39m(genome)\n\u001b[0;32m---> 20\u001b[0m     file \u001b[38;5;241m=\u001b[39m \u001b[43mglob\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mglob\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../../GTDB/all_rep_proteins_aa/*/\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mgenome\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m_protein.faa\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m     21\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m result \u001b[38;5;129;01min\u001b[39;00m SeqIO\u001b[38;5;241m.\u001b[39mparse(file, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfasta\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m     22\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m result\u001b[38;5;241m.\u001b[39mid \u001b[38;5;241m==\u001b[39m genome[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]:\n\u001b[1;32m     23\u001b[0m             \u001b[38;5;66;03m# store seq\u001b[39;00m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "# final-fasta-export.py\n",
    "\n",
    "ars = pd.read_csv('../results/ars_rescheck_nofilt_07212025.csv', index_col=[0,2])\n",
    "\n",
    "arsB_only = pd.read_csv('../results/arsB-only.csv')\n",
    "arsB_only.set_index(['GenomeID', 'Hit'], inplace = True)\n",
    "\n",
    "# export fasta files\n",
    "arsA = ars[(ars.Gene == 'arsA')]\n",
    "arsB = ars[(ars.Gene == 'arsB')]\n",
    "\n",
    "gene_list = [arsA, arsB, arsB_only]\n",
    "gene_names = ['arsA', 'arsB', 'arsB_only']\n",
    "\n",
    "# get fasta sequences for each gene & export to fasta\n",
    "for gene, name in zip(gene_list, gene_names):\n",
    "    print(name)\n",
    "    records = []\n",
    "\n",
    "    for genome,contig in gene.iterrows():\n",
    "        print(genome)\n",
    "        file = glob.glob(f\"../../GTDB/all_rep_proteins_aa/*/{genome[0]}_protein.faa\")[0]\n",
    "        for result in SeqIO.parse(file, \"fasta\"):\n",
    "            if result.id == genome[-1]:\n",
    "                # store seq\n",
    "                gene.loc[genome, 'Seq'] = str(result.seq)\n",
    "                # convert to seqrecord\n",
    "                record = SeqRecord(result.seq, id=genome[0], description=genome[-1])\n",
    "                records.append(record)\n",
    "                # exit loop once sequence is found\n",
    "                break\n",
    "    print(len(records))    \n",
    "    # Write the records to a FASTA file\n",
    "    with open(\"../results/checked_\" + name + \"07212025.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(records, output_handle, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export in genomic-centric organization\n",
    "ars = pd.read_csv('../results/ars_rescheck_nofilt_07212025.csv')\n",
    "ars.reset_index(inplace = True)\n",
    "\n",
    "# load GTDB metadata\n",
    "GTDB_metadata = pd.read_csv('GTDB_metadata.tsv', sep = '\\t', usecols=['accession', 'gtdb_taxonomy', 'ncbi_taxonomy', \n",
    "                                                'ncbi_taxonomy_unfiltered', 'ncbi_country', 'ncbi_isolation_source'])\n",
    "\n",
    "# def for sorting gene order\n",
    "def sort_order(g):\n",
    "    order = ['A', 'B']\n",
    "    return order.index(g)\n",
    "\n",
    "# get gene set (i.e. HDKEN)\n",
    "ars['Gene'] = ars['Gene'].str.replace('ars', '')\n",
    "ars['Gene set'] = ars[['GenomeID','contig','Gene']].groupby(['GenomeID','contig'])['Gene'].transform(lambda x: ''.join(sorted(x, key=sort_order)))\n",
    "# get contig positions (i.e. 1,2,3)\n",
    "ars['index'] = ars['Hit'].str.split('_').str[-1]\n",
    "ars['Position'] = ars[['GenomeID','contig','index']].groupby(['GenomeID','contig'])['index'].transform(lambda x: ','.join(sorted(x)))\n",
    "\n",
    "# group by genome, contig and save\n",
    "ars = ars[['GenomeID', 'contig', 'Gene set', 'Position', 'GTDB']]\n",
    "ars = ars.groupby(['GenomeID','contig']).first()\n",
    "ars.reset_index(level=['GenomeID','contig'], inplace = True)\n",
    "\n",
    "#merge dataframes to include accession, metadata, and sequences (for filtering)\n",
    "ars = pd.merge(ars, GTDB_metadata, left_on = 'GenomeID', right_on = 'accession', how = 'left').drop(columns = ['accession', 'gtdb_taxonomy'])\n",
    "\n",
    "ars['An/aerobic'] = ''\n",
    "ars['Thermophile'] = ''\n",
    "ars['Culture'] = ''\n",
    "ars['Reference'] = ''\n",
    "ars.sort_index(inplace = True)\n",
    "ars.set_index(['GenomeID', 'contig'], inplace = True)\n",
    "\n",
    "#ars.to_csv('ars_genomes_07212025.csv') #--> do not overwrite! contains hand typped info!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "ars.to_csv('ars_genomes_07212025.csv') #--> do not overwrite! contains hand typped info!!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assessing operon organization (what neighbors arsHDKEN clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GB_GCA_000027065.2 FN543096.1\n",
      "GB_GCA_000302695.1 AMFN01000003.1\n",
      "GB_GCA_000752035.1 CCFE01000018.1\n",
      "GB_GCA_000963575.1 CP011077.1\n",
      "GB_GCA_001564065.1 LKNA01000012.1\n",
      "GB_GCA_001636495.1 KV440964.1\n",
      "GB_GCA_001655195.1 LPNY01000021.1\n",
      "GB_GCA_002261155.1 NKXN01000068.1\n",
      "GB_GCA_002400675.1 NVQY01000117.1\n",
      "GB_GCA_002441445.1 DJWC01000119.1\n",
      "GB_GCA_002733685.1 NVWN01000027.1\n",
      "GB_GCA_003457145.1 DMSR01000278.1\n",
      "GB_GCA_003559375.1 PWLN01000067.1\n",
      "GB_GCA_003563455.1 PWRV01000310.1\n",
      "GB_GCA_004153165.1 SEMD01000007.1\n",
      "GB_GCA_004153195.1 SEMC01000021.1\n",
      "GB_GCA_004153225.1 SEMB01000007.1\n",
      "GB_GCA_004195035.1 SGBD01000001.1\n",
      "GB_GCA_004298015.1 SCSE01000033.1\n",
      "GB_GCA_007116585.1 SKIE01000019.1\n",
      "GB_GCA_007123715.1 SKKO01000091.1\n",
      "GB_GCA_007123775.1 SLGK01000236.1\n",
      "GB_GCA_007131625.1 SLQC01000177.1\n",
      "GB_GCA_007132495.1 SLOL01000096.1\n",
      "GB_GCA_007134085.1 SLLL01000035.1\n",
      "GB_GCA_007135335.1 SLJB01000047.1\n",
      "GB_GCA_007693295.1 REDT01000194.1\n",
      "GB_GCA_007693395.1 REDS01000104.1\n",
      "GB_GCA_008180335.1 VTEF01000005.1\n",
      "GB_GCA_009363175.1 CP045205.1\n",
      "GB_GCA_009363175.1 CP045205.1\n",
      "GB_GCA_011077955.1 JAALBX010000021.1\n",
      "GB_GCA_011250595.1 JAANMP010000029.1\n",
      "GB_GCA_013214625.1 JABSPO010000141.1\n",
      "GB_GCA_013215945.1 JABSOM010000020.1\n",
      "GB_GCA_013705725.1 CP055483.1\n",
      "GB_GCA_014595895.1 JABUKB010000001.1\n",
      "GB_GCA_015356505.1 JACYGI010000037.1\n",
      "GB_GCA_015492645.1 WGER01000027.1\n",
      "GB_GCA_015709635.1 CP054180.1\n",
      "GB_GCA_016218575.1 JACRDN010000017.1\n",
      "GB_GCA_016708955.1 JADJES010000024.1\n",
      "GB_GCA_016794365.1 JAEUKB010000033.1\n",
      "GB_GCA_016820585.1 JADHDX010000018.1\n",
      "GB_GCA_018241205.1 JAFEEI010000008.1\n",
      "GB_GCA_018261615.1 JAGTSE010000004.1\n",
      "GB_GCA_018613035.1 JAGGRA010000044.1\n",
      "GB_GCA_018613035.1 JAGGRA010000048.1\n",
      "GB_GCA_019415335.1 DXKH01000137.1\n",
      "GB_GCA_019455325.1 JACFMQ010000001.1\n",
      "GB_GCA_020056205.1 JAGVIF010000149.1\n",
      "GB_GCA_020179385.1 JAIBLJ010000059.1\n",
      "GB_GCA_020524525.2 CP129018.1\n",
      "GB_GCA_020832285.1 JAFIFB010000030.1\n",
      "GB_GCA_020832415.1 JAFIES010000133.1\n",
      "GB_GCA_021366795.1 JAHCPS010000007.1\n",
      "GB_GCA_021774055.1 JAJFJL010000161.1\n",
      "GB_GCA_021777335.1 JAJYDT010000069.1\n",
      "GB_GCA_021786095.1 JAJYNX010000100.1\n",
      "GB_GCA_021786915.1 JAJYMP010000244.1\n",
      "GB_GCA_021790475.1 JAJYDU010000023.1\n",
      "GB_GCA_021791145.1 JAJYXE010000019.1\n",
      "GB_GCA_021791615.1 JAJYWG010000449.1\n",
      "GB_GCA_021793495.1 JAJYSQ010000093.1\n",
      "GB_GCA_021794385.1 JAJYQY010000141.1\n",
      "GB_GCA_021794595.1 JAJYQN010000038.1\n",
      "GB_GCA_021794655.1 JAJYQJ010000112.1\n",
      "GB_GCA_021799615.1 JAJZAC010000049.1\n",
      "GB_GCA_021810205.1 JAJZJC010000016.1\n",
      "GB_GCA_021811865.1 JAKAQB010000054.1\n",
      "GB_GCA_021813195.1 JAKALI010000192.1\n",
      "GB_GCA_021813525.1 JAKAKK010000093.1\n",
      "GB_GCA_021819345.1 JAKARG010000238.1\n",
      "GB_GCA_021821565.1 JAKACI010000030.1\n",
      "GB_GCA_021827365.1 JAKAWK010000076.1\n",
      "GB_GCA_021838635.1 JAKCDC010000054.1\n",
      "GB_GCA_021840555.1 JAKBXW010000003.1\n",
      "GB_GCA_021841005.1 JAKBVZ010000021.1\n",
      "GB_GCA_021850985.1 JAJZJX010000001.1\n",
      "GB_GCA_021907015.1 CP091064.1\n",
      "GB_GCA_022484695.1 JAKVKU010000006.1\n",
      "GB_GCA_023023105.1 CP083825.1\n",
      "GB_GCA_023418355.1 JAEZZU010000069.1\n",
      "GB_GCA_023512345.1 JAILZT010000012.1\n",
      "GB_GCA_023823795.1 CP085710.1\n",
      "GB_GCA_024134725.1 CP099863.1\n",
      "GB_GCA_024228275.1 JAAETO010000089.1\n",
      "GB_GCA_026168805.1 JAPFDS010000013.1\n",
      "GB_GCA_027346035.1 DAIESZ010000074.1\n",
      "GB_GCA_028994555.1 JARDVN010000001.1\n",
      "GB_GCA_029256085.1 JAQSYP010000002.1\n",
      "GB_GCA_029256785.1 JAQSXG010000024.1\n",
      "GB_GCA_030273445.1 CP120678.1\n",
      "GB_GCA_030682075.1 JAUXRH010000188.1\n",
      "GB_GCA_030704225.1 CP131915.1\n",
      "GB_GCA_031180405.1 JAPSLU010000176.1\n",
      "GB_GCA_031181585.1 JAPSKP010000052.1\n",
      "GB_GCA_900112785.1 FONB01000006.1\n",
      "GB_GCA_900455085.1 UGTZ01000002.1\n",
      "GB_GCA_900455155.1 UGUC01000002.1\n",
      "GB_GCA_900696615.1 CAADGS010000012.1\n",
      "GB_GCA_901456055.1 CABDWT010000016.1\n",
      "GB_GCA_901456055.1 CABDWT010000018.1\n",
      "GB_GCA_901472505.1 LR590468.1\n",
      "GB_GCA_913057765.1 CAJXPH010000002.1\n",
      "GB_GCA_913062735.1 CAJXWI010000037.1\n",
      "GB_GCA_937874695.1 CALNAJ010000034.1\n",
      "GB_GCA_937902115.1 CALGSR010000131.1\n",
      "GB_GCA_947497585.1 CANMHG010000001.1\n",
      "GB_GCA_947627865.1 CANQXU010000001.1\n",
      "GB_GCA_947644645.1 CANSAF010000003.1\n",
      "RS_GCF_000010125.1 NC_007351.1\n",
      "RS_GCF_000013705.1 NC_007947.1\n",
      "RS_GCF_000025565.1 NC_014107.1\n",
      "RS_GCF_000025565.1 NC_014121.1\n",
      "RS_GCF_000093085.1 NC_014219.1\n",
      "RS_GCF_000164865.1 NC_014618.1\n",
      "RS_GCF_000171615.1 NZ_ABFU01000056.1\n",
      "RS_GCF_000236785.1 NZ_AGUF01000052.1\n",
      "RS_GCF_000238595.1 NZ_AGTS01000077.1\n",
      "RS_GCF_000253115.1 NC_014145.1\n",
      "RS_GCF_000255495.1 NZ_AHYZ01000008.1\n",
      "RS_GCF_000262305.1 NC_017910.1\n",
      "RS_GCF_000280495.2 NZ_CP018016.1\n",
      "RS_GCF_000285495.1 NZ_HE610977.1\n",
      "RS_GCF_000285495.1 NZ_HE610982.1\n",
      "RS_GCF_000286315.1 NZ_ALJG01000201.1\n",
      "RS_GCF_000299235.1 NC_018649.1\n",
      "RS_GCF_000311725.1 NZ_HE978517.1\n",
      "RS_GCF_000311725.1 NZ_HE978518.1\n",
      "RS_GCF_000311725.1 NZ_HE978518.1\n",
      "RS_GCF_000312485.1 NZ_HE997653.1\n",
      "RS_GCF_000314855.2 NZ_KB233222.1\n",
      "RS_GCF_000377705.1 NZ_KB898632.1\n",
      "RS_GCF_000377985.1 NZ_KB899073.1\n",
      "RS_GCF_000379465.1 NZ_KB905425.1\n",
      "RS_GCF_000379465.1 NZ_KB905434.1\n",
      "RS_GCF_000385175.1 NZ_ARQJ01000020.1\n",
      "RS_GCF_000412335.2 NZ_KE161030.1\n",
      "RS_GCF_000416965.1 NZ_ATCL01000014.1\n",
      "RS_GCF_000427805.1 NZ_ATYS01000001.1\n",
      "RS_GCF_000427825.1 NZ_ATYT01000023.1\n",
      "RS_GCF_000429685.1 NZ_AUCI01000002.1\n",
      "RS_GCF_000463155.2 NZ_AWFZ01000028.1\n",
      "RS_GCF_000474275.2 NZ_ATAE01000033.1\n",
      "RS_GCF_000496755.2 NZ_AYKS02000071.1\n",
      "RS_GCF_000497205.1 NZ_AYJW01000008.1\n",
      "RS_GCF_000497245.1 NZ_AWTC01000004.1\n",
      "RS_GCF_000513095.1 NZ_BAUT01000060.1\n",
      "RS_GCF_000513135.1 NZ_BAUV01000043.1\n",
      "RS_GCF_000517385.1 NZ_BAUW01000006.1\n",
      "RS_GCF_000518885.1 NZ_JAEL01000104.1\n",
      "RS_GCF_000520715.1 NZ_ASSD01000384.1\n",
      "RS_GCF_000520735.1 NZ_ASSC01000369.1\n",
      "RS_GCF_000527095.1 NZ_AZUN01000001.1\n",
      "RS_GCF_000565285.1 NZ_APVL01000006.1\n",
      "RS_GCF_000612505.1 NZ_CP004078.1\n",
      "RS_GCF_000621445.1 NZ_KK211283.1\n",
      "RS_GCF_000648615.1 NZ_JFZC01000041.1\n",
      "RS_GCF_000685865.1 NZ_JMKS01000006.1\n",
      "RS_GCF_000702585.1 NZ_JNIQ01000001.1\n",
      "RS_GCF_000708755.2 NZ_JGVU02000022.1\n",
      "RS_GCF_000733775.1 NZ_JPOE01000002.1\n",
      "RS_GCF_000735375.1 NZ_JMPK01000086.1\n",
      "RS_GCF_000750635.1 NZ_CCDM010000002.1\n",
      "RS_GCF_000756615.1 NZ_CP009288.1\n",
      "RS_GCF_000757885.1 NZ_JQCR01000001.1\n",
      "RS_GCF_000758605.1 NZ_CP009282.1\n",
      "RS_GCF_000761245.1 NZ_AWOR01000066.1\n",
      "RS_GCF_000772945.1 NZ_JPVN01000038.1\n",
      "RS_GCF_000772965.1 NZ_JPVQ01000034.1\n",
      "RS_GCF_000787375.1 NZ_JRJU01000005.1\n",
      "RS_GCF_000813125.1 NZ_BASE01000009.1\n",
      "RS_GCF_000813125.1 NZ_BASE01000083.1\n",
      "RS_GCF_000813345.1 NZ_BBQF01000001.1\n",
      "RS_GCF_000814905.1 NZ_JXAG01000027.1\n",
      "RS_GCF_000829055.1 NZ_AP012544.1\n",
      "RS_GCF_000829395.1 NZ_AP014680.1\n",
      "RS_GCF_000829395.1 NZ_AP014681.1\n",
      "RS_GCF_000829515.1 NZ_JXRR01000001.1\n",
      "RS_GCF_000981805.1 NZ_CP011132.1\n",
      "RS_GCF_000982825.1 NZ_CP011050.1\n",
      "RS_GCF_001037495.1 NZ_KQ089822.1\n",
      "RS_GCF_001076885.1 NZ_JVDT01000219.1\n",
      "RS_GCF_001182045.1 NZ_CVTY01000006.1\n",
      "RS_GCF_001277215.2 NZ_CP013942.1\n",
      "RS_GCF_001293645.1 NZ_LBMD01000007.1\n",
      "RS_GCF_001310895.1 NZ_BAZS01000058.1\n",
      "RS_GCF_001311075.1 NZ_BBAD01000008.1\n",
      "RS_GCF_001311075.1 NZ_BBAD01000096.1\n",
      "RS_GCF_001368795.1 NZ_LN831198.1\n",
      "RS_GCF_001420595.1 NZ_LJIX01000006.1\n",
      "RS_GCF_001428505.1 NZ_LMHU01000001.1\n",
      "RS_GCF_001429185.1 NZ_LMIW01000001.1\n",
      "RS_GCF_001434575.1 NZ_AZDA01000057.1\n",
      "RS_GCF_001434575.1 NZ_AZDA01000098.1\n",
      "RS_GCF_001434935.1 NZ_AZEG01000007.1\n",
      "RS_GCF_001435755.1 NZ_AZFI01000128.1\n",
      "RS_GCF_001436095.1 NZ_AYYK01000009.1\n",
      "RS_GCF_001436295.1 NZ_AYYY01000005.1\n",
      "RS_GCF_001436375.1 NZ_AZGA01000043.1\n",
      "RS_GCF_001436375.1 NZ_AZGA01000066.1\n",
      "RS_GCF_001436395.1 NZ_AZGJ01000005.1\n",
      "RS_GCF_001436505.1 NZ_AYYI01000021.1\n",
      "RS_GCF_001436555.1 NZ_AYZR01000004.1\n",
      "RS_GCF_001457475.1 NZ_LN831029.1\n",
      "RS_GCF_001484765.1 NZ_KQ947385.1\n",
      "RS_GCF_001559075.2 NZ_CP014015.2\n",
      "RS_GCF_001571285.1 NZ_BCTM01000063.1\n",
      "RS_GCF_001591445.1 NZ_BCUX01000056.1\n",
      "RS_GCF_001591645.1 NZ_BCVH01000030.1\n",
      "RS_GCF_001598855.1 NZ_BCZS01000055.1\n",
      "RS_GCF_001598855.1 NZ_BCZS01000063.1\n",
      "RS_GCF_001632845.1 NZ_CP011994.1\n",
      "RS_GCF_001643235.1 NZ_LQWY01000029.1\n",
      "RS_GCF_001655005.1 NZ_LXET01000053.1\n",
      "RS_GCF_001655055.1 NZ_LXEW01000048.1\n",
      "RS_GCF_001721685.1 NZ_CP012502.1\n",
      "RS_GCF_001729745.1 NZ_CP017187.1\n",
      "RS_GCF_001729765.1 NZ_CP017181.1\n",
      "RS_GCF_001730225.1 NZ_MIJE01000031.1\n",
      "RS_GCF_001747895.1 NZ_LNNB01000036.1\n",
      "RS_GCF_001766415.1 NZ_MKXO01000003.1\n",
      "RS_GCF_001870785.1 NZ_MOLV01000027.1\n",
      "RS_GCF_001874625.1 NZ_CP017671.1\n",
      "RS_GCF_001877785.1 NZ_MINN01000096.1\n",
      "RS_GCF_001889165.1 NZ_CP016020.1\n",
      "RS_GCF_001940785.1 NZ_MSLS01000030.1\n",
      "RS_GCF_001955855.1 NZ_MRTS01000016.1\n",
      "RS_GCF_001968835.1 NZ_MRDE01000027.1\n",
      "RS_GCF_001976145.1 NZ_MOXD01000005.1\n",
      "RS_GCF_002019795.1 NZ_KV917381.1\n",
      "RS_GCF_002042885.1 NZ_MVFY01000031.1\n",
      "RS_GCF_002109385.1 NZ_CP020814.1\n",
      "RS_GCF_002153375.1 NZ_MTBQ01000003.1\n",
      "RS_GCF_002156385.1 NZ_MVJM01000002.1\n",
      "RS_GCF_002157855.1 NZ_CP020880.1\n",
      "RS_GCF_002162035.1 NZ_CP021138.1\n",
      "RS_GCF_002162035.1 NZ_CP021139.1\n",
      "RS_GCF_002162315.1 NZ_CP021416.1\n",
      "RS_GCF_002192695.1 NZ_MTLG01000093.1\n",
      "RS_GCF_002209535.1 NZ_NJIG01000012.1\n",
      "RS_GCF_002216775.1 NZ_CP022315.1\n",
      "RS_GCF_002216775.1 NZ_CP022315.1\n",
      "RS_GCF_002252045.1 NZ_NEFF01000004.1\n",
      "RS_GCF_002269255.1 NZ_CP022824.1\n",
      "RS_GCF_002269255.1 NZ_CP022824.1\n",
      "RS_GCF_002706425.1 NZ_CP017713.1\n",
      "RS_GCF_002736105.1 NZ_CP024123.1\n",
      "RS_GCF_002742385.2 NZ_MJBI02000002.1\n",
      "RS_GCF_002797325.1 NZ_KZ454938.1\n",
      "RS_GCF_002797375.1 NZ_KZ454941.1\n",
      "RS_GCF_002797395.1 NZ_KZ454943.1\n",
      "RS_GCF_002797415.1 NZ_KZ454944.1\n",
      "RS_GCF_002843235.3 NZ_CP031123.2\n",
      "RS_GCF_002847015.1 NZ_CP025085.1\n",
      "RS_GCF_002860125.1 NZ_PGVA01000070.1\n",
      "RS_GCF_002860205.1 NZ_PGVB01000018.1\n",
      "RS_GCF_002874535.1 NZ_CP025025.1\n",
      "RS_GCF_002885955.2 NZ_POQS01000004.1\n",
      "RS_GCF_002886255.1 NZ_KZ614149.1\n",
      "RS_GCF_002902305.1 NZ_PPPX01000016.1\n",
      "RS_GCF_002902325.1 NZ_PPPY01000025.1\n",
      "RS_GCF_002902365.1 NZ_PPQC01000019.1\n",
      "RS_GCF_002902565.1 NZ_PPRC01000049.1\n",
      "RS_GCF_002902685.1 NZ_PPRN01000008.1\n",
      "RS_GCF_002970935.1 NZ_CP027194.1\n",
      "RS_GCF_002982135.1 NZ_CP019981.1\n",
      "RS_GCF_002993025.1 NZ_NXGI01000001.1\n",
      "RS_GCF_003014715.1 NZ_PYAV01000009.1\n",
      "RS_GCF_003044065.1 NZ_PZJJ01000025.1\n",
      "RS_GCF_003073375.1 NZ_CP029002.1\n",
      "RS_GCF_003121895.1 NZ_QEWW01000001.1\n",
      "RS_GCF_003144325.1 NZ_PENS01000023.1\n",
      "RS_GCF_003182355.1 NZ_QGTW01000001.1\n",
      "RS_GCF_003184905.1 NZ_KZ821208.1\n",
      "RS_GCF_003184905.1 NZ_QFRU01000012.1\n",
      "RS_GCF_003201565.1 NZ_QJJP01000032.1\n",
      "RS_GCF_003226325.1 NZ_PDOD01000004.1\n",
      "RS_GCF_003259675.1 NZ_PZJG01000002.1\n",
      "RS_GCF_003259695.1 NZ_PZJH01000010.1\n",
      "RS_GCF_003261575.2 NZ_CP036176.1\n",
      "RS_GCF_003315075.1 NZ_QNRJ01000001.1\n",
      "RS_GCF_003316765.1 NZ_NIJF01000045.1\n",
      "RS_GCF_003331145.1 NZ_QLVE01000023.1\n",
      "RS_GCF_003402575.1 NZ_BGZJ01000002.1\n",
      "RS_GCF_003467445.1 NZ_QSGS01000005.1\n",
      "RS_GCF_003570725.1 NZ_QWVS01000013.1\n",
      "RS_GCF_003586445.1 NZ_CP030028.1\n",
      "RS_GCF_003602235.1 NZ_QZVS01000092.1\n",
      "RS_GCF_003610495.1 NZ_RAPK01000008.1\n",
      "RS_GCF_003628445.1 NZ_RBZO01000002.1\n",
      "RS_GCF_003641145.1 NZ_CP032747.1\n",
      "RS_GCF_003718735.1 NZ_CP033460.1\n",
      "RS_GCF_003730295.1 NZ_QTSZ01000041.1\n",
      "RS_GCF_003814815.1 NZ_RKQQ01000002.1\n",
      "RS_GCF_003864415.1 NZ_BFFP01000050.1\n",
      "RS_GCF_003934165.1 NZ_ML133517.1\n",
      "RS_GCF_003937825.1 NZ_RSFW01000014.1\n",
      "RS_GCF_003946115.1 NZ_RHOH01000018.1\n",
      "RS_GCF_003946165.1 NZ_RHOI01000013.1\n",
      "RS_GCF_003946215.1 NZ_RHOW01000013.1\n",
      "RS_GCF_003946245.1 NZ_RHNN01000008.1\n",
      "RS_GCF_003946385.1 NZ_RHOE01000050.1\n",
      "RS_GCF_003946445.1 NZ_RHOF01000071.1\n",
      "RS_GCF_003946515.1 NZ_RHOM01000025.1\n",
      "RS_GCF_003946545.1 NZ_RHOA01000025.1\n",
      "RS_GCF_003946725.1 NZ_RHOY01000019.1\n",
      "RS_GCF_003994415.1 NZ_CP034689.1\n",
      "RS_GCF_004000885.1 NZ_BIMC01000013.1\n",
      "RS_GCF_004005475.1 NZ_RZTZ01000009.1\n",
      "RS_GCF_004006295.1 NZ_CP026095.1\n",
      "RS_GCF_004101865.1 NZ_CP026118.1\n",
      "RS_GCF_004117835.1 NZ_SDSC01000011.1\n",
      "RS_GCF_004124235.1 NZ_CP032487.1\n",
      "RS_GCF_004124235.1 NZ_CP032488.1\n",
      "RS_GCF_004295645.1 NZ_CP034752.1\n",
      "RS_GCF_004336985.1 NZ_SJUO01000002.1\n",
      "RS_GCF_004353845.1 NZ_CP037863.1\n",
      "RS_GCF_004353845.1 NZ_CP037864.1\n",
      "RS_GCF_004358665.1 NZ_SMZR01000002.1\n",
      "RS_GCF_004359575.1 NZ_SCWF01000018.1\n",
      "RS_GCF_004359615.1 NZ_SCWA01000016.1\n",
      "RS_GCF_004367585.1 NZ_CP038015.1\n",
      "RS_GCF_004405125.1 NZ_SORX01000008.1\n",
      "RS_GCF_004684345.1 NZ_CP038469.1\n",
      "RS_GCF_004785665.1 NZ_SRPJ01000005.1\n",
      "RS_GCF_004804275.1 NZ_STGI01000011.1\n",
      "RS_GCF_005217525.1 NZ_PNXQ01000016.1\n",
      "RS_GCF_005404945.1 NZ_BJDG01000001.1\n",
      "RS_GCF_005405165.1 NZ_BJDN01000030.1\n",
      "RS_GCF_005405445.1 NZ_BJEB01000014.1\n",
      "RS_GCF_006539285.1 NZ_BJMS01000064.1\n",
      "RS_GCF_006711645.1 NZ_CP041248.1\n",
      "RS_GCF_006742205.1 NZ_AP019721.1\n",
      "RS_GCF_006861775.1 NZ_VDGG01000024.1\n",
      "RS_GCF_006861825.1 NZ_VDGI01000001.1\n",
      "RS_GCF_006874705.1 NZ_CP035380.1\n",
      "RS_GCF_007115095.1 NZ_SULH01000018.1\n",
      "RS_GCF_007197575.1 NZ_CP041616.1\n",
      "RS_GCF_007830185.1 NZ_VLKZ01000001.1\n",
      "RS_GCF_007830235.1 NZ_VLKI01000009.1\n",
      "RS_GCF_007990205.1 NZ_BJWE01000037.1\n",
      "RS_GCF_007991475.1 NZ_BJYK01000006.1\n",
      "RS_GCF_007994985.1 NZ_VOQF01000008.1\n",
      "RS_GCF_008014775.1 NZ_VPFE01000007.1\n",
      "RS_GCF_008180415.1 NZ_VTEI01000001.1\n",
      "RS_GCF_008180445.1 NZ_VTER01000007.1\n",
      "RS_GCF_008180615.1 NZ_VTEH01000002.1\n",
      "RS_GCF_008180725.1 NZ_VTEV01000003.1\n",
      "RS_GCF_008180865.1 NZ_VTEX01000001.1\n",
      "RS_GCF_008364765.1 NZ_VTUY01000004.1\n",
      "RS_GCF_008369195.1 NZ_VOBC01000001.1\n",
      "RS_GCF_008801815.1 NZ_VZOX01000036.1\n",
      "RS_GCF_008808075.1 NZ_CP019686.1\n",
      "RS_GCF_008932295.1 NZ_WBWA01000042.1\n",
      "RS_GCF_009193155.1 NZ_VUKD01000001.1\n",
      "RS_GCF_009208735.1 NZ_WFLJ01000005.1\n",
      "RS_GCF_009363695.1 NZ_CP045403.1\n",
      "RS_GCF_009664835.1 NZ_CP045929.1\n",
      "RS_GCF_009664975.1 NZ_WJXB01000002.1\n",
      "RS_GCF_009674765.1 NZ_WKKF01000001.1\n",
      "RS_GCF_009674805.1 NZ_WKKI01000020.1\n",
      "RS_GCF_009687905.1 NZ_BLBG01000008.1\n",
      "RS_GCF_009711125.1 NZ_WMIB01000001.1\n",
      "RS_GCF_009720675.1 NZ_WNJO01000030.1\n",
      "RS_GCF_009733865.1 NZ_WOFB01000012.1\n",
      "RS_GCF_009765375.1 NZ_VTOK01000018.1\n",
      "RS_GCF_009857915.1 NZ_WWCM01000028.1\n",
      "RS_GCF_009996735.1 NZ_JAACYI010000001.1\n",
      "RS_GCF_010231475.1 NZ_NRDO01000032.1\n",
      "RS_GCF_011045815.1 NZ_CP047344.1\n",
      "RS_GCF_011046025.1 NZ_CP047639.1\n",
      "RS_GCF_011064845.1 NZ_CP049015.1\n",
      "RS_GCF_011189505.1 NZ_PUJU01000037.1\n",
      "RS_GCF_011398925.1 NZ_AP022842.1\n",
      "RS_GCF_011808225.1 NZ_SOYS01000003.1\n",
      "RS_GCF_012524115.1 NZ_JAAZWB010000066.1\n",
      "RS_GCF_012641075.1 NZ_JAAXLJ010000016.1\n",
      "RS_GCF_012910885.1 NZ_WTVM01000032.1\n",
      "RS_GCF_012911845.1 NZ_JAAXCU010000051.1\n",
      "RS_GCF_012927415.1 NZ_JABBPF010000100.1\n",
      "RS_GCF_012933555.1 NZ_JABBPK010000001.1\n",
      "RS_GCF_013116765.2 NZ_CP074375.1\n",
      "RS_GCF_013177675.1 NZ_CP042652.1\n",
      "RS_GCF_013185195.1 NZ_JAAZWC010000020.1\n",
      "RS_GCF_013283835.1 NZ_CP054052.1\n",
      "RS_GCF_013343715.1 NZ_JABUNR010000002.1\n",
      "RS_GCF_013359935.1 NZ_JABMCB010000169.1\n",
      "RS_GCF_013372265.1 NZ_CP054051.1\n",
      "RS_GCF_013393385.1 NZ_AP023239.1\n",
      "RS_GCF_013394085.1 NZ_CP040910.1\n",
      "RS_GCF_013394855.1 NZ_JABMLK010000002.1\n",
      "RS_GCF_013396335.1 NZ_JABWSY010000014.1\n",
      "RS_GCF_013927005.1 NZ_CP055930.1\n",
      "RS_GCF_013927005.1 NZ_CP055932.1\n",
      "RS_GCF_014169315.1 NZ_AP022133.1\n",
      "RS_GCF_014201095.1 NZ_JACHDN010000001.1\n",
      "RS_GCF_014207655.1 NZ_JACHHJ010000002.1\n",
      "RS_GCF_014217835.1 NZ_CP055263.1\n",
      "RS_GCF_014218705.1 NZ_JABBJF010000047.1\n",
      "RS_GCF_014490035.1 NZ_CP061008.1\n",
      "RS_GCF_014636275.1 NZ_BMEB01000005.1\n",
      "RS_GCF_014636475.1 NZ_BMEL01000005.1\n",
      "RS_GCF_014639435.1 NZ_BMFZ01000001.1\n",
      "RS_GCF_014639435.1 NZ_BMFZ01000004.1\n",
      "RS_GCF_014646815.1 NZ_BMNZ01000003.1\n",
      "RS_GCF_014652395.1 NZ_BMYS01000001.1\n",
      "RS_GCF_014699135.1 NZ_JACOFT010000003.1\n",
      "RS_GCF_014836495.1 NZ_JACSQT010000004.1\n",
      "RS_GCF_014897315.1 NZ_JABUYL010000005.1\n",
      "RS_GCF_014905415.1 NZ_BNEO01000010.1\n",
      "RS_GCF_015355785.1 NZ_JADKYW010000002.1\n",
      "RS_GCF_015672735.1 NZ_JADABY010000005.1\n",
      "RS_GCF_016185115.1 NZ_JACPOF010000004.1\n",
      "RS_GCF_016239305.1 NZ_JAEHOQ010000002.1\n",
      "RS_GCF_016415625.1 NZ_WBXP01000078.1\n",
      "RS_GCF_016526145.1 NZ_JADDUE010000003.1\n",
      "RS_GCF_016526145.1 NZ_JADDUE010000007.1\n",
      "RS_GCF_016785185.1 NZ_JAEKDY010000034.1\n",
      "RS_GCF_016811995.1 NZ_CP069159.1\n",
      "RS_GCF_016820555.1 NZ_JADHDW010000008.1\n",
      "RS_GCF_016879755.1 NZ_WOFU01000001.1\n",
      "RS_GCF_016908395.1 NZ_JAFBDU010000002.1\n",
      "RS_GCF_016908565.1 NZ_JAFBED010000004.1\n",
      "RS_GCF_016908875.1 NZ_JAFBES010000001.1\n",
      "RS_GCF_016909075.1 NZ_JAFBFC010000003.1\n",
      "RS_GCF_017086365.1 NZ_CP070969.1\n",
      "RS_GCF_017303275.1 NZ_JAFKOJ010000007.1\n",
      "RS_GCF_017497975.1 NZ_JAGDEL010000010.1\n",
      "RS_GCF_017569205.1 NZ_CP071870.1\n",
      "RS_GCF_017745675.1 NZ_JAFLJM010000003.1\n",
      "RS_GCF_017815535.1 NZ_CP072596.1\n",
      "RS_GCF_017873715.1 NZ_JAGGKM010000003.1\n",
      "RS_GCF_017921895.1 NZ_JAFIDN010000009.1\n",
      "RS_GCF_017939705.1 NZ_JAGKSQ010000001.1\n",
      "RS_GCF_018075365.1 NZ_CP073011.1\n",
      "RS_GCF_018139565.1 NZ_JAGSPK010000003.1\n",
      "RS_GCF_018252635.1 NZ_JAGTXM010000001.1\n",
      "RS_GCF_018309085.1 NZ_CP070502.1\n",
      "RS_GCF_018323885.1 NZ_AP024585.1\n",
      "RS_GCF_018343525.1 NZ_JAGYPH010000001.1\n",
      "RS_GCF_018424685.1 NZ_QTKX01000003.1\n",
      "RS_GCF_018613155.1 NZ_JAGGQU010000011.1\n",
      "RS_GCF_018613195.1 NZ_JAGGQQ010000009.1\n",
      "RS_GCF_018613195.1 NZ_JAGGQQ010000020.1\n",
      "RS_GCF_018613235.1 NZ_JAGGQP010000029.1\n",
      "RS_GCF_018613315.1 NZ_JAGGQL010000014.1\n",
      "RS_GCF_018613415.1 NZ_JAGGQG010000036.1\n",
      "RS_GCF_018614975.1 NZ_JAGGNR010000007.1\n",
      "RS_GCF_018615055.1 NZ_JAGGNN010000018.1\n",
      "RS_GCF_018617385.1 NZ_JACSKA010000001.1\n",
      "RS_GCF_018617395.1 NZ_JACSKC010000001.1\n",
      "RS_GCF_018617855.1 NZ_JACSKY010000012.1\n",
      "RS_GCF_018918075.1 NZ_JAHLPW010000008.1\n",
      "RS_GCF_018918075.1 NZ_JAHLPW010000011.1\n",
      "RS_GCF_018966525.1 NZ_CP076607.1\n",
      "RS_GCF_019024285.1 NZ_WMKZ01000002.1\n",
      "RS_GCF_019039105.1 NZ_JAHPIV010000006.1\n",
      "RS_GCF_019048065.1 NZ_CP077291.1\n",
      "RS_GCF_019237635.1 NZ_JAFAGS010000069.1\n",
      "RS_GCF_019242955.1 NZ_CP066129.1\n",
      "RS_GCF_019243775.1 NZ_CP069157.1\n",
      "RS_GCF_019343475.1 NZ_CP072453.1\n",
      "RS_GCF_019599185.1 NZ_JAIEUL010000008.1\n",
      "RS_GCF_019599185.1 NZ_JAIEUL010000020.1\n",
      "RS_GCF_019655335.1 NZ_BPFD01000052.1\n",
      "RS_GCF_019749255.1 NZ_JACWFH010000035.1\n",
      "RS_GCF_019752415.1 NZ_JAIEME010000073.1\n",
      "RS_GCF_019837165.1 NZ_JAGYHC010000001.1\n",
      "RS_GCF_019931755.1 NZ_CP082918.1\n",
      "RS_GCF_020075705.1 NZ_JAIQUM010000083.1\n",
      "RS_GCF_020171655.1 NZ_JAIVAC010000001.1\n",
      "RS_GCF_020172085.1 NZ_JAIVAT010000007.1\n",
      "RS_GCF_020381325.1 NZ_JAGSPI010000006.1\n",
      "RS_GCF_020519665.1 NZ_CP084703.1\n",
      "RS_GCF_020524325.1 NZ_JAIVKY010000002.1\n",
      "RS_GCF_020524345.1 NZ_JAIVLH010000019.1\n",
      "RS_GCF_020731355.1 NZ_JAJFZK010000039.1\n",
      "RS_GCF_020734105.1 NZ_JAJAFR010000004.1\n",
      "RS_GCF_020832205.1 NZ_JAFIFD010000026.1\n",
      "RS_GCF_021030645.1 NZ_JACGCJ010000040.1\n",
      "RS_GCF_021560035.1 NZ_CP074373.1\n",
      "RS_GCF_021991925.1 NZ_RXZI01000003.1\n",
      "RS_GCF_022049065.2 NZ_CP095746.1\n",
      "RS_GCF_022170825.1 NZ_CP081878.1\n",
      "RS_GCF_022171985.1 NZ_JAJVHF010000020.1\n",
      "RS_GCF_022427965.1 NZ_JAKTTI010000041.1\n",
      "RS_GCF_022484265.1 NZ_JAKVLR010000001.1\n",
      "RS_GCF_022812285.1 NZ_CP093300.1\n",
      "RS_GCF_022819085.1 NZ_JALHAG010000001.1\n",
      "RS_GCF_022835675.1 NZ_BQUT01000001.1\n",
      "RS_GCF_022919835.1 NZ_CP095074.1\n",
      "RS_GCF_022921115.1 NZ_CP095075.1\n",
      "RS_GCF_023100685.1 NZ_CP090064.1\n",
      "RS_GCF_023168165.1 NZ_BQYJ01000017.1\n",
      "RS_GCF_023168345.1 NZ_AP025689.1\n",
      "RS_GCF_023205015.1 NZ_CP096258.1\n",
      "RS_GCF_023283765.1 NZ_JAKILG010000007.1\n",
      "RS_GCF_023515995.1 NZ_CP097320.1\n",
      "RS_GCF_023522835.1 NZ_JALCZU010000001.1\n",
      "RS_GCF_023555415.1 NZ_JAKRYL010000013.1\n",
      "RS_GCF_023573145.1 NZ_CP097638.1\n",
      "RS_GCF_023715205.1 NZ_JAMBNR010000011.1\n",
      "RS_GCF_023715225.1 NZ_JAMBNS010000010.1\n",
      "RS_GCF_023715485.1 NZ_JAMBOF010000007.1\n",
      "RS_GCF_023715705.1 NZ_JAMBOS010000037.1\n",
      "RS_GCF_024023655.1 NZ_JAJJVQ010000006.1\n",
      "RS_GCF_024189295.1 NZ_JAIZEN010000008.1\n",
      "RS_GCF_024189315.1 NZ_JAIZEO010000006.1\n",
      "RS_GCF_024267655.1 NZ_CP100595.1\n",
      "RS_GCF_024362205.1 NZ_CP101462.1\n",
      "RS_GCF_024385605.1 NZ_JANFOJ010000008.1\n",
      "RS_GCF_024609785.1 NZ_JANHJQ010000001.1\n",
      "RS_GCF_024718525.1 NZ_JAMXAZ010000042.1\n",
      "RS_GCF_025234635.1 NZ_JANIEG010000010.1\n",
      "RS_GCF_025234665.1 NZ_JANIEH010000001.1\n",
      "RS_GCF_025234715.1 NZ_JANIEI010000023.1\n",
      "RS_GCF_025234735.1 NZ_JANIEL010000059.1\n",
      "RS_GCF_025234775.1 NZ_JANIEK010000035.1\n",
      "RS_GCF_025398035.1 NZ_CP104558.1\n",
      "RS_GCF_026011715.1 NZ_BRVM01000005.1\n",
      "RS_GCF_026013365.1 NZ_JAPDOE010000007.1\n",
      "RS_GCF_026547225.1 NZ_CP106878.1\n",
      "RS_GCF_026659735.1 NZ_JANSKK010000008.1\n",
      "RS_GCF_026798195.1 NZ_JAPQEZ010000001.1\n",
      "RS_GCF_027186265.1 NZ_CP114176.1\n",
      "RS_GCF_027322205.1 NZ_DAHVHY010000075.1\n",
      "RS_GCF_028069725.1 NZ_JANEWG010000053.1\n",
      "RS_GCF_028610675.1 NZ_JAQQWS010000001.1\n",
      "RS_GCF_028610705.1 NZ_JAQQWT010000006.1\n",
      "RS_GCF_028736235.1 NZ_CP117884.1\n",
      "RS_GCF_028765345.1 NZ_JARAJZ010000026.1\n",
      "RS_GCF_028970245.1 NZ_JAPEVP010000009.1\n",
      "RS_GCF_029024945.1 NZ_CP118977.1\n",
      "RS_GCF_029028005.1 NZ_JAOTPO010000002.1\n",
      "RS_GCF_029094545.1 NZ_CP117835.1\n",
      "RS_GCF_029318635.1 NZ_CP119676.1\n",
      "RS_GCF_029319185.1 NZ_CP119754.1\n",
      "RS_GCF_029667115.1 NZ_JARUIU010000004.1\n",
      "RS_GCF_029714185.1 NZ_CP121671.1\n",
      "RS_GCF_029717145.1 NZ_CP099222.1\n",
      "RS_GCF_029891575.1 NZ_JARZHF010000005.1\n",
      "RS_GCF_030053835.1 NZ_CP125655.1\n",
      "RS_GCF_030053895.1 NZ_CP125781.1\n",
      "RS_GCF_030137545.1 NZ_JARUVL010000025.1\n",
      "RS_GCF_030332225.1 NZ_JAOWIF010000035.1\n",
      "RS_GCF_030345335.1 NZ_JAPVRC010000016.1\n",
      "RS_GCF_030400225.1 NZ_VJYB01000011.1\n",
      "RS_GCF_030433715.1 NZ_JAROCC010000007.1\n",
      "RS_GCF_030651495.1 NZ_JAUSFI010000092.1\n",
      "RS_GCF_030676415.1 NZ_JAUUTW010000009.1\n",
      "RS_GCF_030704685.1 NZ_JAUZDX010000003.1\n",
      "RS_GCF_030812995.1 NZ_JAUSTT010000006.1\n",
      "RS_GCF_030813355.1 NZ_JAUSUD010000021.1\n",
      "RS_GCF_030813365.1 NZ_JAUSUC010000003.1\n",
      "RS_GCF_030813635.1 NZ_JAUSUM010000001.1\n",
      "RS_GCF_030817055.1 NZ_JAUSYV010000001.1\n",
      "RS_GCF_030908545.1 NZ_CP133083.1\n",
      "RS_GCF_031181385.1 NZ_JAPSKV010000021.1\n",
      "RS_GCF_900088825.1 NZ_LT592170.1\n",
      "RS_GCF_900088825.1 NZ_LT592171.1\n",
      "RS_GCF_900089495.1 NZ_LT605203.1\n",
      "RS_GCF_900089495.1 NZ_LT605203.1\n",
      "RS_GCF_900094795.1 NZ_FMBC01000015.1\n",
      "RS_GCF_900094925.1 NZ_FMAY01000008.1\n",
      "RS_GCF_900099605.1 NZ_FNDK01000016.1\n",
      "RS_GCF_900099765.1 NZ_FNDX01000002.1\n",
      "RS_GCF_900103695.1 NZ_FNHF01000005.1\n",
      "RS_GCF_900104185.1 NZ_FNIZ01000025.1\n",
      "RS_GCF_900104555.1 NZ_FNJU01000015.1\n",
      "RS_GCF_900106555.1 NZ_FNOE01000010.1\n",
      "RS_GCF_900108305.1 NZ_FNUX01000044.1\n",
      "RS_GCF_900109925.1 NZ_FOBW01000009.1\n",
      "RS_GCF_900109925.1 NZ_FOBW01000021.1\n",
      "RS_GCF_900110815.1 NZ_FOIT01000007.1\n",
      "RS_GCF_900112475.1 NZ_FOLW01000017.1\n",
      "RS_GCF_900113325.1 NZ_FOOY01000016.1\n",
      "RS_GCF_900114445.1 NZ_FOTB01000005.1\n",
      "RS_GCF_900114745.1 NZ_FOUB01000069.1\n",
      "RS_GCF_900115925.1 NZ_FOYF01000007.1\n",
      "RS_GCF_900128965.1 NZ_FQUL01000059.1\n",
      "RS_GCF_900162695.1 NZ_FUIH01000007.1\n",
      "RS_GCF_900166625.1 NZ_LT800499.1\n",
      "RS_GCF_900166665.1 NZ_LT800494.1\n",
      "RS_GCF_900168685.1 NZ_FVZN01000012.1\n",
      "RS_GCF_900200035.1 NZ_FZTC01000042.1\n",
      "RS_GCF_900206265.1 NZ_LT907782.1\n",
      "RS_GCF_900240995.1 NZ_LT965954.1\n",
      "RS_GCF_900635455.1 NZ_LR134116.1\n",
      "RS_GCF_900635495.1 NZ_LR134136.1\n",
      "RS_GCF_902459665.1 NZ_CABPSL010000004.1\n",
      "RS_GCF_902809865.1 NZ_CADDWH010000001.1\n",
      "RS_GCF_903645295.1 NZ_CAHJXM010000003.1\n",
      "RS_GCF_943737015.1 NZ_CALSFP010000001.1\n",
      "RS_GCF_945906775.1 NZ_CAMDRE010000002.1\n",
      "RS_GCF_945906775.1 NZ_CAMDRE010000003.1\n"
     ]
    }
   ],
   "source": [
    "# operon organization\n",
    "\n",
    "from cluster_pos import cluster_pos\n",
    "\n",
    "# grab checked ars\n",
    "ars = pd.read_csv('../results/ars_rescheck_nofilt_07212025.csv')\n",
    "\n",
    "#multi-index to cluster by genome, contig\n",
    "ars.reset_index(inplace = True)\n",
    "ars.set_index(['GenomeID', 'contig'], inplace = True)\n",
    "ars.sort_index(inplace = True)\n",
    "ars.drop_duplicates(inplace = True)\n",
    "\n",
    "# iterate through each genome and contig\n",
    "for genome in ars.index.get_level_values(0).unique(): # iterate through each genome\n",
    "    for contig in ars.loc[genome].index.get_level_values(0).unique(): # iterate through each contig\n",
    "\n",
    "        tmp = ars.loc[(genome, contig)]\n",
    "\n",
    "        # get clusters\n",
    "        pos_clusters = cluster_pos(tmp.Hit.unique(), 40)\n",
    "        # get positions within +/-12 genes of center \n",
    "        for cl in pos_clusters:\n",
    "            middle = cl[len(cl)//2] # center of cluster\n",
    "            pos = [middle+num for num in range(-12,13) if middle+num > 0]\n",
    "            acc = [contig + '_' + str(p) for p in pos] # get acc\n",
    "\n",
    "            # write list items into a .txt file\n",
    "            os.system(\"rm seqs.txt\")\n",
    "            with open(\"seqs.txt\", \"w\") as f:\n",
    "                for item in acc:\n",
    "                    f.write(f\"{item}\\n\")\n",
    "\n",
    "            # save subsets as fasta\n",
    "            try:\n",
    "                file = glob.glob(f\"../../GTDB/all_rep_proteins_aa/*/{genome[0]}_protein.faa\")[0]\n",
    "                os.system(f\"seqtk subseq {file} seqs.txt > ../operon_org/input-fastas/{genome}_{contig}_operon.fasta\")\n",
    "            except:\n",
    "                continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot operon-org of annotated genes\n",
    "import requests\n",
    "\n",
    "arsAB = pd.read_csv('../results/ars_rescheck_nofilt_07212025.csv', index_col=[0,1])\n",
    "\n",
    "gene_abv = {'K00532': 'hydA', 'K07486':'tns', 'K07485':'tns', 'K07487':'tns', 'K13819': 'arsU'} # store ko2gene\n",
    "gene_data = pd.DataFrame(columns = ['genome', 'contig', 'query_id', 'gene', 'ko_number', 'start', 'end', 'orientation'])\n",
    "#gene_data.set_index(['query_id'], inplace = True)\n",
    "\n",
    "def ko2gene(ko):\n",
    "    try:\n",
    "        return gene_abv[ko]\n",
    "    except:\n",
    "        url = f\"https://rest.kegg.jp/get/ko:{ko}\"\n",
    "        r = requests.get(url)\n",
    "        for line in r.text.split(\"\\n\"):\n",
    "            if line.startswith(\"SYMBOL\"):\n",
    "                gene = line.split()[-1]\n",
    "                gene_abv[ko] = gene\n",
    "                return gene\n",
    "\n",
    "# grab annotation files\n",
    "for file in glob.glob(f\"../operon-org/microbeannotator/annotation_results/*.annot\")[0:5]:\n",
    "    # convet ko_number to gene abv\n",
    "    annot = pd.read_csv(file, sep = '\\t', index_col=0)\n",
    "    annot['gene'] = annot['ko_number'].apply(ko2gene)\n",
    "\n",
    "    # grab genome and contig from file name\n",
    "    s = r'([A-Z_]+[A-Z0-9.]+)_([A-Z_]*[A-Z0-9.]+)'\n",
    "    genome = re.search(s, file).group(1)\n",
    "    contig = re.search(s, file).group(2)\n",
    "\n",
    "    # grab start, end, orientation of each gene\n",
    "    input = glob.glob(f\"../operon-org/input-fastas/{genome}_{contig}_operon.fasta\")[0]\n",
    "    for result in SeqIO.parse(input, \"fasta\"):\n",
    "        start = int(result.description.split('# ')[1])\n",
    "        end = int(result.description.split('# ')[2])\n",
    "        orientation = int(result.description.split('# ')[3])\n",
    "        if result.id in annot.index:\n",
    "            annot.loc[result.id, 'start'] = start\n",
    "            annot.loc[result.id, 'end'] = end\n",
    "            annot.loc[result.id, 'orientation'] = orientation\n",
    "\n",
    "    # add arsA, arsB hits to annot\n",
    "    for row in arsAB.loc[(genome, contig)].iterrows():\n",
    "        annot.loc[row[1].Hit, 'gene'] = row[1].Gene\n",
    "        annot.loc[row[1].Hit, 'start'] = int(row[1].Location.split('-')[0])\n",
    "        annot.loc[row[1].Hit, 'end'] = int(row[1].Location.split('-')[1])\n",
    "        annot.loc[row[1].Hit, 'orientation'] = 1 # BUG:CANNOT KNOW ORIENTATION\n",
    "\n",
    "    # set genome and contig\n",
    "    annot['genome'] = genome\n",
    "    annot['contig'] = contig\n",
    "\n",
    "    annot.reset_index(inplace = True)\n",
    "    gene_data = pd.concat([gene_data, annot[['genome', 'contig', 'query_id', 'gene', 'ko_number', 'start', 'end', 'orientation']]])\n",
    "gene_data.to_csv('../operon-org/operon-org-plot-data-test.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot\n",
    "from pygenomeviz import GenomeViz\n",
    "\n",
    "# open microbeannotator results\n",
    "gene_data = pd.read_csv('../operon-org/operon-org-plot-data-test.csv', index_col=[1,2])\n",
    "genome_list = gene_data.index.get_level_values(0).unique().to_list()\n",
    "\n",
    "gv = GenomeViz(feature_track_ratio=0.7)\n",
    "gv.set_scale_bar(labelsize=20)\n",
    "# gv.set_scale_xticks(labelsize=20)\n",
    "\n",
    "for genome in genome_list:\n",
    "    genes = gene_data.loc[(genome)].gene.to_list()\n",
    "    starts = gene_data.loc[(genome)].start.to_list()\n",
    "    ends = gene_data.loc[(genome)].end.to_list()\n",
    "    orientations = gene_data.loc[(genome)].orientation.to_list()\n",
    "\n",
    "    track = gv.add_feature_track(genome, (int(min(starts)), int(max(ends))))\n",
    "    for idx, gene in enumerate(genes):\n",
    "        if gene == 'arsA':\n",
    "            color = '#156082'\n",
    "        elif gene == 'arsR':\n",
    "            color = '#76C0AF'\n",
    "        elif gene == 'arsC':\n",
    "            color = '#98A2C2'\n",
    "        elif gene == 'arsB':\n",
    "            color = '#C85D2B'\n",
    "        elif gene == 'arsD':\n",
    "            color = 'orange'\n",
    "        else:\n",
    "            color = '#FFFFFF'\n",
    "            \n",
    "        track.add_feature(\n",
    "            int(starts[idx]),\n",
    "            int(ends[idx]),\n",
    "            int(orientations[idx]),\n",
    "            plotstyle='bigarrow',\n",
    "            arrow_shaft_ratio=0.7,\n",
    "            fc = color,\n",
    "            lw = 1,\n",
    "            label = gene,\n",
    "            text_kws=dict(rotation=0, vpos=\"bottom\", hpos=\"center\"))\n",
    "        \n",
    "    #track.add_sublabel()#int(min(starts)), int(max(ends))), fontsize=20)\n",
    "\n",
    "gv.savefig(\"../operon-org/test.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OLD Below"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read fasta file merged_arsH.faa and save as df\n",
    "\n",
    "# arsH = list(SeqIO.parse(\"clustered_arsH_rep_seq.fasta\", \"fasta\"))\n",
    "# arsH = pd.DataFrame([(record.id, str(record.seq)) for record in arsH], columns = ['GenomeID', 'Seq'])\n",
    "\n",
    "arsD_list = list(SeqIO.parse(\"ko_arsD.fasta\", \"fasta\"))\n",
    "arsD_list = pd.DataFrame([(record.id, str(record.seq)) for record in arsD_list], columns = ['GenomeID', 'Seq'])\n",
    "arsD_list = arsD_list.GenomeID\n",
    "\n",
    "# arsD = list(SeqIO.parse(\"clustered_arsD_rep_seq.fasta\", \"fasta\"))\n",
    "# arsD = pd.DataFrame([(record.id, str(record.seq)) for record in arsD], columns = ['GenomeID', 'Seq'])\n",
    "\n",
    "# arsK = list(SeqIO.parse(\"clustered_arsK_rep_seq.fasta\", \"fasta\"))\n",
    "# arsK = pd.DataFrame([(record.id, str(record.seq)) for record in arsK], columns = ['GenomeID', 'Seq'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# final\n",
    "# CDD conserved features and error tolerance\n",
    "\n",
    "arsD_FeMoCo = (r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{79,81}VGG[A-Z]R[A-Z]{21,23}F[A-Z]{60}([A-Z]{50,52})?H', 5)\n",
    "# to reduce computation time, remove non-indentity residues (except Cys as it is involved in holding MoFe) from the pattern\n",
    "arsD_FeMoCo_mod = (r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{106,110}F[A-Z]{60}([A-Z]{50,52})?H', 10)\n",
    "# for all other arsD check their length!\n",
    "arsD_Pcluster = (r'C[A-Z]Y[A-Z]{20}P[A-Z]GC[A-Z]{64,65}C[A-Z]{30,31}G', 5)\n",
    "\n",
    "arsK_Pcluster = (r'C[A-Z]P[A-Z]{19}S[A-Z]GC[A-Z]{2}YF[A-Z]{51,53}TC[A-Z]{30,35}S', 7)\n",
    "\n",
    "arsH_active = (r'KGGIGKS[A-Z]{24,36}KAD', 4)\n",
    "arsH_FeS = (r'C[A-Z]{29,37}C[A-Z]G',2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "typical nifD done\n",
      "nifD done\n"
     ]
    }
   ],
   "source": [
    "# check seeds seq\n",
    "\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import numpy as np\n",
    "\n",
    "from check_seq import check_seq\n",
    "from check_seq import conditional_check_seq\n",
    "\n",
    "#arsH\n",
    "\n",
    "# #pass constant arguments to check_seq\n",
    "# feature = arsH_active\n",
    "# filled_check_seq = partial(check_seq, feature = feature)\n",
    "# # parallelize iterables with Pool\n",
    "# with Pool(4) as p:\n",
    "#     results = p.map(filled_check_seq, arsH.Seq)\n",
    "\n",
    "# arsH.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# # match FeS cluster binding site (binds Fe4-S4 cluster)\n",
    "# feature = arsH_FeS\n",
    "# filled_check_seq = partial(check_seq, feature = feature)\n",
    "# # parallelize iterables with Pool\n",
    "# with Pool(4) as p:\n",
    "#     results = p.map(filled_check_seq, arsH.Seq)\n",
    "# arsH.loc[:, 'match2'] = np.array(results)\n",
    "\n",
    "# print('arsH done')\n",
    "\n",
    "# match arsD MoFeCo binding site\n",
    "feature = arsD_FeMoCo\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, arsD.Seq)\n",
    "arsD.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# match arsD MoFeCo binding site\n",
    "feature = arsD_Pcluster\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, arsD.Seq)\n",
    "arsD.loc[:, 'match2'] = np.array(results)\n",
    "\n",
    "print('typical arsD done')\n",
    "\n",
    "# match arsD MoFeCo binding site\n",
    "feature = arsD_FeMoCo\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "\n",
    "# Use Pool to apply the function\n",
    "with Pool(4) as p:\n",
    "    results = p.starmap(conditional_check_seq, zip(arsD.Seq, arsD.match1))\n",
    "\n",
    "# match arsD MoFeCo binding site\n",
    "feature = arsD_FeMoCo_mod\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, arsD.Seq)\n",
    "arsD.match2 = results\n",
    "\n",
    "print('arsD done')\n",
    "\n",
    "# # match arsD MoFeCo binding site --> all arsK pass\n",
    "# feature = arsK_Pcluster\n",
    "# filled_check_seq = partial(check_seq, feature = feature)\n",
    "# # parallelize iterables with Pool\n",
    "# with Pool(4) as p:\n",
    "#     results = p.map(filled_check_seq, arsK.Seq)\n",
    "\n",
    "\n",
    "# arsK.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# print('arsK done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<regex.Match object; span=(73, 192), match='WGPIKDMIHISHGPVGCGYWSWSGRRNYYIGTTGIDTFGTMHFTSDFQERDIVFGGDKKLVKLIQELDVLFPLNRGVSIQSECPIGLIGDDIEAVARKTSKEIGKPVVPVRCEGFRGVS', fuzzy_counts=(4, 0, 1)>\n"
     ]
    }
   ],
   "source": [
    "# print rows where arsH.match1 is false\n",
    "# at 5, all but 2 rows fail rsin:B6N60_04474, dou:BMF77_01967\n",
    "# failed_arsH = arsH.loc[arsH.match1 != True]\n",
    "\n",
    "# failed = list(failed_arsH.Seq)\n",
    "\n",
    "arsH_active = (r'KGGIGKS[A-Z]{24,36}KAD', 5)\n",
    "\n",
    "arsK_Pcluster = (r'C[A-Z]P[A-Z]{19}S[A-Z]GC[A-Z]{2}YF[A-Z]{51,53}TC[A-Z]{30,35}S', 6)\n",
    "\n",
    "arsD_FeMoCo = (r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{79,81}VGG[A-Z]R[A-Z]{21,23}F[A-Z]{60}([A-Z]{50,52})?H', 6)\n",
    "# to reduce computation time, remove non-indentity residues (except Cys as it is involved in holding MoFe) from the pattern\n",
    "arsD_FeMoCo_mod = (r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{106,110}F[A-Z]{60}([A-Z]{50,52})?H', 10)\n",
    "\n",
    "seq = 'MTPPENQNIIEERKELIKEVLDAYPDKAKKKREKHLSVYEEGKSDCGVKSNIKSLPGVMTARGCAYAGSKGVVWGPIKDMIHISHGPVGCGYWSWSGRRNYYIGTTGIDTFGTMHFTSDFQERDIVFGGDKKLVKLIQELDVLFPLNRGVSIQSECPIGLIGDDIEAVARKTSKEIGKPVVPVRCEGFRGVSQSLGHHIANDMVRDWVFTRADQAKKDGTLQFEGTPYDVAIIGDYNIGGDAWASRILLEEIGLRVVAQWSGDGTINEMLMTPNVKMNLIHCYRSMNYISRHMEEAYGIPCLNYDNVTMTDENIRQIAFYGKGGIGKSTTSQNTLAAMAEMGQRILIVGCDPKADSTRLMLHSKAQTTVLHLAAERGAVEDIEIEEVMLTGFRNVRCVESGGPEPGVGCAGRGIITAINFLEENGAYQDLDFVSYDVLGDVVCGGFAMPIREGKAQEIYIVTSGEMMAMYAANNIARGVLKYAHTGGVRLGGLICNSRNTDREIELIETLAKRLNTQMIHYVPRDNIVQHAELRRMTVNEYAPESNQANEYRILAQKIIDNKNLAIPTPIEMEELEELLIEFGILESDENAAKLVGKTAAEAPVS'\n",
    "\n",
    "fuzzy_pattern = f'({arsK_Pcluster[0]}){{e<={arsK_Pcluster[1]}}}'\n",
    "\n",
    "match = regex.search(fuzzy_pattern, seq, regex.BESTMATCH)\n",
    "\n",
    "print(match)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nifH done\n",
      "typical nifD done\n",
      "nifK done\n"
     ]
    }
   ],
   "source": [
    "# similar to final but in python file\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import numpy as np\n",
    "\n",
    "from check_seq import check_seq\n",
    "\n",
    "# arsH\n",
    "\n",
    "# pass constant arguments to check_seq\n",
    "feature = arsH_active\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, arsH.Seq)\n",
    "\n",
    "arsH.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# match FeS cluster binding site (binds Fe4-S4 cluster)\n",
    "feature = arsH_FeS\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, arsH.Seq)\n",
    "arsH.loc[:, 'match2'] = np.array(results)\n",
    "\n",
    "print('arsH done')\n",
    "\n",
    "# match arsD MoFeCo binding site\n",
    "feature = arsD_FeMoCo\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, arsD.Seq)\n",
    "arsD.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# match arsD MoFeCo binding site\n",
    "feature = arsD_Pcluster\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, arsD.Seq)\n",
    "arsD.loc[:, 'match2'] = np.array(results)\n",
    "\n",
    "print('typical arsD done')\n",
    "\n",
    "# match arsD MoFeCo binding site\n",
    "feature = arsD_FeMoCo\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "\n",
    "# Use Pool to apply the function\n",
    "with Pool(4) as p:\n",
    "    results = p.starmap(conditional_check_seq, zip(arsD.Seq, arsD.match1))\n",
    "\n",
    "# match arsD MoFeCo binding site\n",
    "feature = arsD_FeMoCo_mod\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, arsD.Seq)\n",
    "arsD.match2 = results\n",
    "\n",
    "print('arsD done')\n",
    "\n",
    "# match arsD MoFeCo binding site\n",
    "feature = arsK_Pcluster\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, arsK.Seq)\n",
    "\n",
    "\n",
    "arsK.loc[:, 'match1'] = np.array(results)\n",
    "arsK = arsK.drop(columns = 'match2')\n",
    "\n",
    "print('arsK done')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What do I do with arsH and arsD seq that fail CDD criteria? (usually they are too short and only contain a portion of the gene)?\n",
    "    - instead of hmmsearch with all NFixDB onlny use initial seeds seq?\n",
    "    - def remove them from arsH (sequences that are long enough dont have correct residues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# finalized checks, e such that all seqs in clustered_ars_seq.fasta pass -- KEEP!!!\n",
    "\n",
    "arsD_FeMoCo = r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{79,81}VGG[A-Z]R[A-Z]{21,23}F[A-Z]{60}([A-Z]{50,52})?H' # e<=4\n",
    "# to reduce computation time, remove non-indentity residues (except Cys as it is involved in holding MoFe) from the pattern\n",
    "arsD_FeMoCo_mod = r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{106,110}F[A-Z]{60}([A-Z]{50,52})?H' #e<=8\n",
    "# for all other arsD check their length!\n",
    "\n",
    "arsD_Pcluster = r'C[A-Z]Y[A-Z]{20}P[A-Z]GC[A-Z]{64,65}C[A-Z]{30,31}G' # e<=5\n",
    "\n",
    "arsK_Pcluster = r'C[A-Z]P[A-Z]{19}S[A-Z]GC[A-Z]{2}YF[A-Z]{51,53}TC[A-Z]{30,35}S' # e<=8\n",
    "\n",
    "arsH_active = r'KGGIGKS[A-Z]{24,36}KAD' # e<=2\n",
    "arsH_FeS = r'C[A-Z]{29,37}C[A-Z]G' # e<=2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "# grab sequences from same genomes and string together before clustering\n",
    "\n",
    "# join df on genomeID\n",
    "\n",
    "ars = arsH.merge(arsD, on = 'GenomeID', how = 'inner', suffixes = ('_H', '_D'))\n",
    "ars = ars.merge(arsK, on = 'GenomeID', how = 'inner', suffixes = ('_K', '_K'))\n",
    "ars['Full Seq'] = ars.Seq_H + ars.Seq_D + ars.Seq\n",
    "\n",
    "\n",
    "# Convert the DataFrame to SeqRecord objects\n",
    "records = []\n",
    "for index, row in ars.iterrows():\n",
    "    record = SeqRecord(Seq(row['Full Seq']), id=index, description=\"arsHDK\")\n",
    "    records.append(record)\n",
    "\n",
    "# Write the records to a FASTA file\n",
    "with open(\"filteredhits.fasta\", \"w\") as output_handle:\n",
    "    SeqIO.write(records, output_handle, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now I need to cluster the hits by sequence similarity before predicting their structures\n",
    "file = 'clustered_filteredhits_30_rep_seq.fasta'\n",
    "\n",
    "# get list of ids from fasta file\n",
    "ids = []\n",
    "for record in SeqIO.parse(file, \"fasta\"):\n",
    "    ids.append(record.id)\n",
    "\n",
    "clustered_ars = ars[ars.index.isin(ids)]\n",
    "\n",
    "clustered_ars = clustered_ars[['Seq_H', 'Seq_D', 'Seq']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra functions!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phylum\n",
      "p__Pseudomonadota      7029\n",
      "p__Bacillota_A         5453\n",
      "p__Desulfobacterota    2859\n",
      "p__Bacteroidota        1311\n",
      "p__Cyanobacteriota     1198\n",
      "                       ... \n",
      "p__Nitrospirota_B         3\n",
      "p__JACRDZ01               3\n",
      "p__SZUA-182               3\n",
      "p__UBA3054                3\n",
      "p__Asgardarchaeota        3\n",
      "Name: count, Length: 62, dtype: int64\n",
      "class\n",
      "c__Clostridia             5383\n",
      "c__Gammaproteobacteria    3961\n",
      "c__Alphaproteobacteria    2983\n",
      "c__Cyanobacteriia         1190\n",
      "c__Bacteroidia             861\n",
      "                          ... \n",
      "c__Limnochordia              3\n",
      "c__Methanomethylicia         3\n",
      "c__SURF-8                    3\n",
      "c__SZUA-365                  3\n",
      "c__JACRHL01                  3\n",
      "Name: count, Length: 144, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# taxa counter\n",
    "ars = pd.read_feather('ars_final_04202025.feather')\n",
    "ars['phylum'] = ars['GTDB'].str.split(';').str[1]\n",
    "ars['class'] = ars['GTDB'].str.split(';').str[2]\n",
    "ars['order'] = ars['GTDB'].str.split(';').str[3]\n",
    "ars['family'] = ars['GTDB'].str.split(';').str[4]\n",
    "\n",
    "phyla = ars['phylum'].value_counts()\n",
    "print(phyla)\n",
    "\n",
    "classes = ars['class'].value_counts()\n",
    "print(classes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clean up newick file nodes = Species | Gene | GenomeID | Hit\n",
    "import Bio.Phylo.NewickIO\n",
    "\n",
    "def replace_char(string):\n",
    "    # Replace the unwanted characters with an empty string\n",
    "    return string.replace(' ', '_').replace(':', '_').replace('(', '_').replace(')', '_').replace(',', '_').replace(';', '_')\n",
    "\n",
    "# export in genomic-centric organization\n",
    "ars = pd.read_feather('ars_final_04292025.feather')\n",
    "ars.reset_index(inplace = True)\n",
    "ars.set_index(['GenomeID', 'Hit'], inplace = True)\n",
    "\n",
    "# supposedly prevents decimal truncation by bio.phlyo...but doesn't work\n",
    "Bio.Phylo.NewickIO._format_branch_length = lambda bl: f\"{bl:.10f}\" if bl is not None else ''\n",
    "\n",
    "# load GTDB metadata --> comment out bc it takes a while to load\n",
    "GTDB_metadata = pd.read_csv('GTDB_metadata.tsv', sep = '\\t', usecols=['accession', 'gtdb_taxonomy', 'ncbi_taxonomy', \n",
    "                                                'ncbi_taxonomy_unfiltered', 'ncbi_country', 'ncbi_isolation_source'])\n",
    "\n",
    "# change names from treefiles to only be GTDB genomes\n",
    "for gene in 'DK':\n",
    "    file = f'../trees/ars{gene}_noOut_04292025/clustered_ars{gene}_noOut_rep_seq.trim.treefile'\n",
    "    tree = Phylo.read(file, \"newick\")\n",
    "    envs = []\n",
    "    # Extract and print node names\n",
    "    for clade in tree.get_terminals():\n",
    "        if clade.name.startswith('WP_'):\n",
    "            new_name = clade.name\n",
    "        else:\n",
    "            genome = clade.name.split('.')[0] + '.' + clade.name.split('.')[1][0]\n",
    "            hit = '.'.join(clade.name.split('.')[1:])[2:]\n",
    "            species = ars.loc[(genome, hit), 'GTDB'].split('__')[-1]\n",
    "            gene_name = ars.loc[(genome, hit), 'Gene']\n",
    "            env = GTDB_metadata.loc[GTDB_metadata.accession == genome, 'ncbi_isolation_source'].values[0]\n",
    "            env = replace_char(env)\n",
    "            new_name = f\"{species}|{gene_name}|{env}|{genome}|{hit}\"\n",
    "        clade.name = new_name\n",
    "\n",
    "    # Save the modified tree to a new file\n",
    "    Phylo.write(tree, f'../trees/ars{gene}_noOut_04292025/ars{gene}_env.treefile', \"newick\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get putative diazotroph genome names from LAMG results \n",
    "filteredhits_LAMG = list(SeqIO.parse(\"filteredhits_LAMG.fasta\", \"fasta\"))\n",
    "diazotrophs = [rec.name for rec in filteredhits_LAMG] \n",
    "\n",
    "# save as text file\n",
    "with open(\"diazotrophs_LAMG.txt\", \"w\") as f:\n",
    "    for item in diazotrophs:\n",
    "        f.write(\"%s\\n\" % item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# grab LAMG arsH seq \n",
    "arsH = pd.read_feather('filtered_arsH_LAMG.feather')\n",
    "\n",
    "# Convert the DataFrame to SeqRecord objects\n",
    "records = []\n",
    "for index, row in arsH.iterrows():\n",
    "    record = SeqRecord(Seq(row['Seq_H']), id=index, description=\"arsH_hmm\")\n",
    "    records.append(record)\n",
    "\n",
    "# Write the records to a FASTA file\n",
    "with open(\"filtered_arsH_LAMG.fasta\", \"w\") as output_handle:\n",
    "    SeqIO.write(records, output_handle, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([  1.,   0.,   0.,   0.,   0.,   0.,   1.,   0.,   0.,   1.,   1.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   3.,   0.,   1.,\n",
       "          0.,   0.,   1.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   1.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   1.,   5.,   1.,   5.,\n",
       "          5.,   2.,   1.,   1., 148.,  77., 178.,  82.,  54.,  40.,  25.,\n",
       "         11.,  11.,   9.,   2.,   0.,   2.,   2.,   2.,   1.,   1.,   1.,\n",
       "          2.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   1.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   2.]),\n",
       " array([438.   , 440.045, 442.09 , 444.135, 446.18 , 448.225, 450.27 ,\n",
       "        452.315, 454.36 , 456.405, 458.45 , 460.495, 462.54 , 464.585,\n",
       "        466.63 , 468.675, 470.72 , 472.765, 474.81 , 476.855, 478.9  ,\n",
       "        480.945, 482.99 , 485.035, 487.08 , 489.125, 491.17 , 493.215,\n",
       "        495.26 , 497.305, 499.35 , 501.395, 503.44 , 505.485, 507.53 ,\n",
       "        509.575, 511.62 , 513.665, 515.71 , 517.755, 519.8  , 521.845,\n",
       "        523.89 , 525.935, 527.98 , 530.025, 532.07 , 534.115, 536.16 ,\n",
       "        538.205, 540.25 , 542.295, 544.34 , 546.385, 548.43 , 550.475,\n",
       "        552.52 , 554.565, 556.61 , 558.655, 560.7  , 562.745, 564.79 ,\n",
       "        566.835, 568.88 , 570.925, 572.97 , 575.015, 577.06 , 579.105,\n",
       "        581.15 , 583.195, 585.24 , 587.285, 589.33 , 591.375, 593.42 ,\n",
       "        595.465, 597.51 , 599.555, 601.6  , 603.645, 605.69 , 607.735,\n",
       "        609.78 , 611.825, 613.87 , 615.915, 617.96 , 620.005, 622.05 ,\n",
       "        624.095, 626.14 , 628.185, 630.23 , 632.275, 634.32 , 636.365,\n",
       "        638.41 , 640.455, 642.5  , 644.545, 646.59 , 648.635, 650.68 ,\n",
       "        652.725, 654.77 , 656.815, 658.86 , 660.905, 662.95 , 664.995,\n",
       "        667.04 , 669.085, 671.13 , 673.175, 675.22 , 677.265, 679.31 ,\n",
       "        681.355, 683.4  , 685.445, 687.49 , 689.535, 691.58 , 693.625,\n",
       "        695.67 , 697.715, 699.76 , 701.805, 703.85 , 705.895, 707.94 ,\n",
       "        709.985, 712.03 , 714.075, 716.12 , 718.165, 720.21 , 722.255,\n",
       "        724.3  , 726.345, 728.39 , 730.435, 732.48 , 734.525, 736.57 ,\n",
       "        738.615, 740.66 , 742.705, 744.75 , 746.795, 748.84 , 750.885,\n",
       "        752.93 , 754.975, 757.02 , 759.065, 761.11 , 763.155, 765.2  ,\n",
       "        767.245, 769.29 , 771.335, 773.38 , 775.425, 777.47 , 779.515,\n",
       "        781.56 , 783.605, 785.65 , 787.695, 789.74 , 791.785, 793.83 ,\n",
       "        795.875, 797.92 , 799.965, 802.01 , 804.055, 806.1  , 808.145,\n",
       "        810.19 , 812.235, 814.28 , 816.325, 818.37 , 820.415, 822.46 ,\n",
       "        824.505, 826.55 , 828.595, 830.64 , 832.685, 834.73 , 836.775,\n",
       "        838.82 , 840.865, 842.91 , 844.955, 847.   ]),\n",
       " <BarContainer object of 200 artists>)"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAigAAAGdCAYAAAA44ojeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8hTgPZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAjE0lEQVR4nO3df2zV1eH/8de1P66AbbXU9t6rpVQDcaMd8mMT8QctIlIrOOsQ/FkCkhkZkwET0I+hGKVEo+IkMjWIIBCImTK1RiwqoEMnlmEKGixSoGhrM4ReCngL9Hz/WHh/ubRQCvdyz7s+H8k74b5/9ZwdWZ/c3tvrMcYYAQAAWOS8WA8AAADgRAQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOvEx3oAZ6K5uVk//PCDkpKS5PF4Yj0cAABwGowx2r9/vwKBgM4779TPkbgyUH744QdlZmbGehgAAOAM1NTU6NJLLz3lOa4MlKSkJEn/m2BycnKMRwMAAE5HMBhUZmam8338VFwZKMd+rJOcnEygAADgMqfz8gxeJAsAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKICluk8vU/fpZbEeBgDEBIECAACs0+5AWbdunYYPH65AICCPx6OVK1eGHfd4PK1uTz/9tHNOXl5ei+OjR48+68kAAICOod2BcuDAAfXu3Vvz5s1r9XhtbW3Y9uqrr8rj8ej2228PO2/8+PFh57300ktnNgMAANDhxLf3goKCAhUUFJz0uM/nC3v8z3/+U/n5+brsssvC9nfu3LnFuQAAAFKUX4Py448/qqysTOPGjWtxbOnSpUpLS1OvXr00depU7d+//6T3CYVCCgaDYRsAAOi42v0MSnssWrRISUlJKioqCtt/9913Kzs7Wz6fT5s3b9aMGTP01Vdfqby8vNX7lJaWatasWdEcKgAAsEhUA+XVV1/V3XffrfPPPz9s//jx450/5+TkqEePHurfv782btyovn37trjPjBkzNHnyZOdxMBhUZmZm9AYOAABiKmqB8sknn2jr1q1asWJFm+f27dtXCQkJqqqqajVQvF6vvF5vNIYJAAAsFLXXoCxYsED9+vVT79692zx3y5YtOnz4sPx+f7SGAwAAXKTdz6A0NjZq27ZtzuPq6mpt2rRJqamp6tatm6T//QjmjTfe0DPPPNPi+u+++05Lly7VzTffrLS0NH399deaMmWK+vTpo2uuueYspgIAADqKdgfKl19+qfz8fOfxsdeGFBcX67XXXpMkLV++XMYY3XnnnS2uT0xM1Icffqjnn39ejY2NyszMVGFhoWbOnKm4uLgznAYAAOhIPMYYE+tBtFcwGFRKSooaGhqUnJwc6+EAUXHsc3h2zCmM8UgAIDLa8/2bz+IBAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB12h0o69at0/DhwxUIBOTxeLRy5cqw42PGjJHH4wnbBgwYEHZOKBTSxIkTlZaWpi5dumjEiBHavXv3WU0EAAB0HO0OlAMHDqh3796aN2/eSc8ZNmyYamtrne29994LOz5p0iS99dZbWr58uT799FM1Njbqlltu0dGjR9s/AwAA0OHEt/eCgoICFRQUnPIcr9crn8/X6rGGhgYtWLBAr7/+uoYMGSJJWrJkiTIzM7V69WrddNNN7R0SAADoYKLyGpQ1a9YoPT1dPXv21Pjx41VfX+8cq6io0OHDhzV06FBnXyAQUE5OjtavXx+N4QAAAJdp9zMobSkoKNDIkSOVlZWl6upqPfbYYxo8eLAqKirk9XpVV1enxMREXXTRRWHXZWRkqK6urtV7hkIhhUIh53EwGIz0sAEAgEUiHiijRo1y/pyTk6P+/fsrKytLZWVlKioqOul1xhh5PJ5Wj5WWlmrWrFmRHioAALBU1N9m7Pf7lZWVpaqqKkmSz+dTU1OT9u7dG3ZefX29MjIyWr3HjBkz1NDQ4Gw1NTXRHjYAAIihqAfKnj17VFNTI7/fL0nq16+fEhISVF5e7pxTW1urzZs3a+DAga3ew+v1Kjk5OWwDAAAdV7t/xNPY2Kht27Y5j6urq7Vp0yalpqYqNTVVJSUluv322+X3+7Vjxw498sgjSktL02233SZJSklJ0bhx4zRlyhR17dpVqampmjp1qnJzc5139QAAgF+2dgfKl19+qfz8fOfx5MmTJUnFxcWaP3++KisrtXjxYu3bt09+v1/5+flasWKFkpKSnGuee+45xcfH64477tChQ4d0ww036LXXXlNcXFwEpgS4U/fpZZKkHXMKYzwSAIi9dgdKXl6ejDEnPb5q1ao273H++efrhRde0AsvvNDeLw8AAH4B+CweAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWaXegrFu3TsOHD1cgEJDH49HKlSudY4cPH9a0adOUm5urLl26KBAI6L777tMPP/wQdo+8vDx5PJ6wbfTo0Wc9GQAA0DG0O1AOHDig3r17a968eS2OHTx4UBs3btRjjz2mjRs36s0339S3336rESNGtDh3/Pjxqq2tdbaXXnrpzGYAAAA6nPj2XlBQUKCCgoJWj6WkpKi8vDxs3wsvvKDf/e532rVrl7p16+bs79y5s3w+X3u/PAAA+AWI+mtQGhoa5PF4dOGFF4btX7p0qdLS0tSrVy9NnTpV+/fvP+k9QqGQgsFg2AYAADqudj+D0h4///yzpk+frrvuukvJycnO/rvvvlvZ2dny+XzavHmzZsyYoa+++qrFsy/HlJaWatasWdEcKgAAsEjUAuXw4cMaPXq0mpub9eKLL4YdGz9+vPPnnJwc9ejRQ/3799fGjRvVt2/fFveaMWOGJk+e7DwOBoPKzMyM1tABAECMRSVQDh8+rDvuuEPV1dX66KOPwp49aU3fvn2VkJCgqqqqVgPF6/XK6/VGY6gAAMBCEQ+UY3FSVVWljz/+WF27dm3zmi1btujw4cPy+/2RHg4AAHChdgdKY2Ojtm3b5jyurq7Wpk2blJqaqkAgoD/84Q/auHGj3n33XR09elR1dXWSpNTUVCUmJuq7777T0qVLdfPNNystLU1ff/21pkyZoj59+uiaa66J3MwAAIBrtTtQvvzyS+Xn5zuPj702pLi4WCUlJXr77bclSVdeeWXYdR9//LHy8vKUmJioDz/8UM8//7waGxuVmZmpwsJCzZw5U3FxcWcxFQAA0FG0O1Dy8vJkjDnp8VMdk6TMzEytXbu2vV8WAAD8gvBZPAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsE67A2XdunUaPny4AoGAPB6PVq5cGXbcGKOSkhIFAgF16tRJeXl52rJlS9g5oVBIEydOVFpamrp06aIRI0Zo9+7dZzURAADQcbQ7UA4cOKDevXtr3rx5rR5/6qmn9Oyzz2revHnasGGDfD6fbrzxRu3fv985Z9KkSXrrrbe0fPlyffrpp2psbNQtt9yio0ePnvlMAABAhxHf3gsKCgpUUFDQ6jFjjObOnatHH31URUVFkqRFixYpIyNDy5Yt0x//+Ec1NDRowYIFev311zVkyBBJ0pIlS5SZmanVq1frpptuOovpAACAjiCir0Gprq5WXV2dhg4d6uzzer0aNGiQ1q9fL0mqqKjQ4cOHw84JBALKyclxzgEAAL9s7X4G5VTq6uokSRkZGWH7MzIytHPnTuecxMREXXTRRS3OOXb9iUKhkEKhkPM4GAxGctgAAMAyUXkXj8fjCXtsjGmx70SnOqe0tFQpKSnOlpmZGbGxAgAA+0Q0UHw+nyS1eCakvr7eeVbF5/OpqalJe/fuPek5J5oxY4YaGhqcraamJpLDBgAAlolooGRnZ8vn86m8vNzZ19TUpLVr12rgwIGSpH79+ikhISHsnNraWm3evNk550Rer1fJyclhGwAA6Lja/RqUxsZGbdu2zXlcXV2tTZs2KTU1Vd26ddOkSZM0e/Zs9ejRQz169NDs2bPVuXNn3XXXXZKklJQUjRs3TlOmTFHXrl2VmpqqqVOnKjc313lXDwAA+GVrd6B8+eWXys/Pdx5PnjxZklRcXKzXXntNDz/8sA4dOqQHH3xQe/fu1VVXXaUPPvhASUlJzjXPPfec4uPjdccdd+jQoUO64YYb9NprrykuLi4CUwIAAG7nMcaYWA+ivYLBoFJSUtTQ0MCPe9BhdJ9eJknaMaew1ccA4Hbt+f7NZ/EAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOvEx3oAAE6t+/Qy58875hTGcCQAcO7wDAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwTsQDpXv37vJ4PC22CRMmSJLGjBnT4tiAAQMiPQzAtbpPLwv77bEA8EsU8V91v2HDBh09etR5vHnzZt14440aOXKks2/YsGFauHCh8zgxMTHSwwAAAC4W8UC5+OKLwx7PmTNHl19+uQYNGuTs83q98vl8kf7SAACgg4jqa1Campq0ZMkSjR07Vh6Px9m/Zs0apaenq2fPnho/frzq6+tPeZ9QKKRgMBi2AQCAjiuqgbJy5Urt27dPY8aMcfYVFBRo6dKl+uijj/TMM89ow4YNGjx4sEKh0EnvU1paqpSUFGfLzMyM5rABAECMeYwxJlo3v+mmm5SYmKh33nnnpOfU1tYqKytLy5cvV1FRUavnhEKhsIAJBoPKzMxUQ0ODkpOTIz5uIBZO54WxO+YUnoORAEB0BINBpaSknNb374i/BuWYnTt3avXq1XrzzTdPeZ7f71dWVpaqqqpOeo7X65XX6430EAEAgKWi9iOehQsXKj09XYWFp/4X3549e1RTUyO/3x+toQAAAJeJSqA0Nzdr4cKFKi4uVnz8/3+SprGxUVOnTtVnn32mHTt2aM2aNRo+fLjS0tJ02223RWMoAADAhaLyI57Vq1dr165dGjt2bNj+uLg4VVZWavHixdq3b5/8fr/y8/O1YsUKJSUlRWMoAADAhaISKEOHDlVrr73t1KmTVq1aFY0vCQAAOhA+iwcAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAF+k+vUzdp5fFehgAEHUECgAAsA6BAgAArEOgAAAA60Q8UEpKSuTxeMI2n8/nHDfGqKSkRIFAQJ06dVJeXp62bNkS6WEAAAAXi8ozKL169VJtba2zVVZWOseeeuopPfvss5o3b542bNggn8+nG2+8Ufv374/GUAAAgAtFJVDi4+Pl8/mc7eKLL5b0v2dP5s6dq0cffVRFRUXKycnRokWLdPDgQS1btiwaQwEAAC4UlUCpqqpSIBBQdna2Ro8ere3bt0uSqqurVVdXp6FDhzrner1eDRo0SOvXrz/p/UKhkILBYNgGAAA6rogHylVXXaXFixdr1apVeuWVV1RXV6eBAwdqz549qqurkyRlZGSEXZORkeEca01paalSUlKcLTMzM9LDBgAAFol4oBQUFOj2229Xbm6uhgwZorKy//1SqUWLFjnneDyesGuMMS32HW/GjBlqaGhwtpqamkgPGwAAWCTqbzPu0qWLcnNzVVVV5byb58RnS+rr61s8q3I8r9er5OTksA0AAHRcUQ+UUCikb775Rn6/X9nZ2fL5fCovL3eONzU1ae3atRo4cGC0hwIAAFwiPtI3nDp1qoYPH65u3bqpvr5eTzzxhILBoIqLi+XxeDRp0iTNnj1bPXr0UI8ePTR79mx17txZd911V6SHAgAAXCrigbJ7927deeed+u9//6uLL75YAwYM0Oeff66srCxJ0sMPP6xDhw7pwQcf1N69e3XVVVfpgw8+UFJSUqSHAgAAXMpjjDGxHkR7BYNBpaSkqKGhgdejoMNoz6cU75hTGMWRAEB0tOf7N5/FAwAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAC7UfXpZuz79GADchkABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWiXiglJaW6re//a2SkpKUnp6u3//+99q6dWvYOWPGjJHH4wnbBgwYEOmhAAAAl4p4oKxdu1YTJkzQ559/rvLych05ckRDhw7VgQMHws4bNmyYamtrne29996L9FAAAIBLxUf6hu+//37Y44ULFyo9PV0VFRW6/vrrnf1er1c+ny/SXx4AAHQAUX8NSkNDgyQpNTU1bP+aNWuUnp6unj17avz48aqvrz/pPUKhkILBYNgGAAA6Lo8xxkTr5sYY3Xrrrdq7d68++eQTZ/+KFSt0wQUXKCsrS9XV1Xrsscd05MgRVVRUyOv1trhPSUmJZs2a1WJ/Q0ODkpOTozV84JzqPr3sjK/dMacwgiMBgOgIBoNKSUk5re/fUQ2UCRMmqKysTJ9++qkuvfTSk55XW1urrKwsLV++XEVFRS2Oh0IhhUIh53EwGFRmZiaBgg6FQAHQ0bUnUCL+GpRjJk6cqLffflvr1q07ZZxIkt/vV1ZWlqqqqlo97vV6W31mBQAAdEwRDxRjjCZOnKi33npLa9asUXZ2dpvX7NmzRzU1NfL7/ZEeDgAAcKGIv0h2woQJWrJkiZYtW6akpCTV1dWprq5Ohw4dkiQ1NjZq6tSp+uyzz7Rjxw6tWbNGw4cPV1pamm677bZIDwcAALhQxJ9BmT9/viQpLy8vbP/ChQs1ZswYxcXFqbKyUosXL9a+ffvk9/uVn5+vFStWKCkpKdLDAQAALhSVH/GcSqdOnbRq1apIf1kAANCB8Fk8AADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDrxsR4AgLPXfXpZi3075hTGYCQAEBk8gwIAAKxDoAAAAOsQKEAH1X16Was/+gEANyBQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHX3UPRMix3zly7FfM8+vnAeDM8QwKAACwDoECAACsQ6AAAADrECgux+etAAA6IgIFAABYJ6bv4nnxxRf19NNPq7a2Vr169dLcuXN13XXXxXJIQFSd+E6fc/k1j/+6p3rW7WRja+0+ADqeWPz/VGti9gzKihUrNGnSJD366KP6z3/+o+uuu04FBQXatWtXrIYEAAAsEbNnUJ599lmNGzdO999/vyRp7ty5WrVqlebPn6/S0tJYDUvSmdXjqf51eS5/H8apxm5LFUdarOd14vqezmuCbH7dkM1jwy9HrP9eI/ZiEihNTU2qqKjQ9OnTw/YPHTpU69evb3F+KBRSKBRyHjc0NEiSgsFgVMbXHDrY7vsfu6a1644/drJzztSJYz3V2M9kXm4Q63m1tr62OvG/k7O9DxAtsf57/UsWzf/tj93TGNP2ySYGvv/+eyPJ/Otf/wrb/+STT5qePXu2OH/mzJlGEhsbGxsbG1sH2GpqatpshZi+SNbj8YQ9Nsa02CdJM2bM0OTJk53Hzc3N+umnn9S1a9dWzz/XgsGgMjMzVVNTo+Tk5FgPJ6KYmzt15LlJHXt+zM2dmNvpMcZo//79CgQCbZ4bk0BJS0tTXFyc6urqwvbX19crIyOjxfler1derzds34UXXhjNIZ6R5OTkDvcf5jHMzZ068tykjj0/5uZOzK1tKSkpp3VeTN7Fk5iYqH79+qm8vDxsf3l5uQYOHBiLIQEAAIvE7Ec8kydP1r333qv+/fvr6quv1ssvv6xdu3bpgQceiNWQAACAJWIWKKNGjdKePXv0+OOPq7a2Vjk5OXrvvfeUlZUVqyGdMa/Xq5kzZ7b4MVRHwNzcqSPPTerY82Nu7sTcIs9jzOm81wcAAODc4bN4AACAdQgUAABgHQIFAABYh0ABAADWIVBOQ2lpqTwejyZNmuTsGzNmjDweT9g2YMCAsOtCoZAmTpyotLQ0denSRSNGjNDu3bvP8ehbKikpaTF2n8/nHDfGqKSkRIFAQJ06dVJeXp62bNkSdg+3zs3N6yZJ33//ve655x517dpVnTt31pVXXqmKigrnuJvXrq25uXntunfv3mLsHo9HEyZMkOTudWtrbm5etyNHjuj//u//lJ2drU6dOumyyy7T448/rubmZucct67d6cwt5mt31h+s08F98cUXpnv37uY3v/mNeeihh5z9xcXFZtiwYaa2ttbZ9uzZE3btAw88YC655BJTXl5uNm7caPLz803v3r3NkSNHzvEsws2cOdP06tUrbOz19fXO8Tlz5pikpCTzj3/8w1RWVppRo0YZv99vgsGgc45b5+bmdfvpp59MVlaWGTNmjPn3v/9tqqurzerVq822bducc9y6dqczNzevXX19fdi4y8vLjSTz8ccfG2Pcu27GtD03N6/bE088Ybp27WreffddU11dbd544w1zwQUXmLlz5zrnuHXtTmdusV47AuUU9u/fb3r06GHKy8vNoEGDWgTKrbfeetJr9+3bZxISEszy5cudfd9//70577zzzPvvvx/FUbdt5syZpnfv3q0ea25uNj6fz8yZM8fZ9/PPP5uUlBTz97//3Rjj3rkZ4+51mzZtmrn22mtPetzNa9fW3Ixx99qd6KGHHjKXX365aW5udvW6teb4uRnj7nUrLCw0Y8eODdtXVFRk7rnnHmOMu//OtTU3Y2K/dvyI5xQmTJigwsJCDRkypNXja9asUXp6unr27Knx48ervr7eOVZRUaHDhw9r6NChzr5AIKCcnBytX78+6mNvS1VVlQKBgLKzszV69Ght375dklRdXa26urqwcXu9Xg0aNMgZt1vndoxb1+3tt99W//79NXLkSKWnp6tPnz565ZVXnONuXru25naMW9fueE1NTVqyZInGjh0rj8fj6nU70YlzO8at63bttdfqww8/1LfffitJ+uqrr/Tpp5/q5ptvluTuv3Ntze2YWK5dTD/N2GbLly/Xxo0btWHDhlaPFxQUaOTIkcrKylJ1dbUee+wxDR48WBUVFfJ6vaqrq1NiYqIuuuiisOsyMjJafEjiuXbVVVdp8eLF6tmzp3788Uc98cQTGjhwoLZs2eKM7cQPbczIyNDOnTslybVz69q1q6vXbfv27Zo/f74mT56sRx55RF988YX+/Oc/y+v16r777nP12rU1N8ndf+eOt3LlSu3bt09jxoyRJFev24lOnJvk7nWbNm2aGhoadMUVVyguLk5Hjx7Vk08+qTvvvFOSu9eurblJsV87AqUVNTU1euihh/TBBx/o/PPPb/WcUaNGOX/OyclR//79lZWVpbKyMhUVFZ303saYsH9ZxEJBQYHz59zcXF199dW6/PLLtWjRIucFUCeO8XTGbfvcJk+e7Op1a25uVv/+/TV79mxJUp8+fbRlyxbNnz/f+SYuuXPtTmdubl674y1YsEAFBQUtPm7ejet2otbm5uZ1W7FihZYsWaJly5apV69e2rRpkyZNmqRAIKDi4mLnPDeu3enMLdZrx494WlFRUaH6+nr169dP8fHxio+P19q1a/W3v/1N8fHxOnr0aItr/H6/srKyVFVVJUny+XxqamrS3r17w86rr69vUdux1qVLF+Xm5qqqqsp5x8uJ9Xv8uN06t9a4ad38fr9+/etfh+371a9+pV27dkmSq9eurbmd7Bq3rN0xO3fu1OrVq3X//fc7+9y8bsdrbW6tcdO6/fWvf9X06dM1evRo5ebm6t5779Vf/vIXlZaWSnL32rU1t9ac67UjUFpxww03qLKyUps2bXK2/v376+6779amTZsUFxfX4po9e/aopqZGfr9fktSvXz8lJCSovLzcOae2tlabN2/WwIEDz9lcTkcoFNI333wjv9+v7Oxs+Xy+sHE3NTVp7dq1zrjdOrfWuGndrrnmGm3dujVs37fffut8wKab166tubXGTWt3zMKFC5Wenq7CwkJnn5vX7Xitza01blq3gwcP6rzzwr9NxsXFOW/FdfPatTW31pzztTvrl9n+Qhz/Lp79+/ebKVOmmPXr15vq6mrz8ccfm6uvvtpccsklLd5adumll5rVq1ebjRs3msGDB8f8rWXGGDNlyhSzZs0as337dvP555+bW265xSQlJZkdO3YYY/73trmUlBTz5ptvmsrKSnPnnXe2+rY5t83N7ev2xRdfmPj4ePPkk0+aqqoqs3TpUtO5c2ezZMkS5xy3rl1bc3P72hljzNGjR023bt3MtGnTWhxz67odc7K5uX3diouLzSWXXOK8FffNN980aWlp5uGHH3bOcevatTU3G9aOQDlNxwfKwYMHzdChQ83FF19sEhISTLdu3UxxcbHZtWtX2DWHDh0yf/rTn0xqaqrp1KmTueWWW1qcEwvH3qefkJBgAoGAKSoqMlu2bHGONzc3m5kzZxqfz2e8Xq+5/vrrTWVlZdg93Dg3t6+bMca88847Jicnx3i9XnPFFVeYl19+Oey4m9fuVHPrCGu3atUqI8ls3bq1xTE3r5sxJ5+b29ctGAyahx56yHTr1s2cf/755rLLLjOPPvqoCYVCzjluXbu25mbD2nmMMebsn4cBAACIHF6DAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsM7/A9HoypdGX4xXAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "\n",
    "# how many hits have short aln len?\n",
    "inputs = SeqIO.parse('../Ars-input/arsA_C113_172_422.fasta', 'fasta')\n",
    "lens = [len(rec.seq) for rec in inputs]\n",
    "plt.hist(lens, bins=200)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>gtdb_taxonomy</th>\n",
       "      <th>ncbi_country</th>\n",
       "      <th>ncbi_isolation_source</th>\n",
       "      <th>ncbi_taxonomy</th>\n",
       "      <th>ncbi_taxonomy_unfiltered</th>\n",
       "      <th>Seq_nifD</th>\n",
       "      <th>Seq_nifK</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GB_GCA_903822005.1</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>none</td>\n",
       "      <td>lake water</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...</td>\n",
       "      <td>MSTAIDDTRKIIDEVLEAYPEKVRKARARHLAPNDPAGSCASCKIQ...</td>\n",
       "      <td>MSKLKIKDHSELFLEPEYVDLFERKKEFEDACPDEEVKRVLDWTKT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GB_GCA_937889015.1</td>\n",
       "      <td>d__Bacteria;p__Planctomycetota;c__Planctomycet...</td>\n",
       "      <td>none</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>d__Bacteria;p__Planctomycetota;c__Planctomycet...</td>\n",
       "      <td>d__Bacteria;x__PVC group;p__Planctomycetota;c_...</td>\n",
       "      <td>MAPELIDPERVKQELLQAYPAKVGRKRAGQMVVNRKDTPGESPEIV...</td>\n",
       "      <td>MNLLKGAVRGSQENPTVLLRHTTPNVVERQALTINPAKTCQPIGAM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GB_GCA_937924475.1</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>none</td>\n",
       "      <td>Mariane sediment</td>\n",
       "      <td>d__Bacteria;p__;c__;o__;f__;g__;s__</td>\n",
       "      <td>d__Bacteria;x__environmental samples;s__uncult...</td>\n",
       "      <td>MNTYTINKKNEEEIRHHRVCRSRGGESCAFDGAMIVLQPIADAVHL...</td>\n",
       "      <td>MFAISTTCMAEVIGDDLNAFIKKAKQQGVIPEQLPTPYSNTSSFTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GB_GCA_002339325.1</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>South Africa:Carltonville</td>\n",
       "      <td>water</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...</td>\n",
       "      <td>MTRIIEQDTQRIVDEVLEVYPEKTKKERKKHLLANDPECSSSTCGL...</td>\n",
       "      <td>MKVQDHVKLYRREDYEKLFRDKREFECPYPEEEVKKVEEYTKTEEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GB_GCA_020725335.1</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>Russia: Borehole near Beyelii Yar in Tomsk</td>\n",
       "      <td>groundwater filtered on 0.22 m filters</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>MPRIIEQDTEKLIEEVLEIYPEKTRKERRKHLISNDPECTTSTCQL...</td>\n",
       "      <td>MKIMDHVELYKSEDYKELFENKKQFECPYSEEEVKKIEEYTKTEEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1978</th>\n",
       "      <td>GB_GCA_026418275.1</td>\n",
       "      <td>d__Bacteria;p__Bacillota_A;c__Clostridia;o__Ac...</td>\n",
       "      <td>Malaysia: Sungkai; Perak</td>\n",
       "      <td>alkaline hot spring water</td>\n",
       "      <td>d__Bacteria;p__Bacillota;c__Clostridia;o__;f__...</td>\n",
       "      <td>d__Bacteria;x__Terrabacteria group;p__Bacillot...</td>\n",
       "      <td>MPINIQNIEAATRENRLGAITGYHGTIKDLVSKATCCSLKNRERCF...</td>\n",
       "      <td>MTQLIEQLRHVCSLGALQSVLAIHRAVPILHAGPGCGQKLWSATGV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>GB_GCA_026419705.1</td>\n",
       "      <td>d__Bacteria;p__Cyanobacteriota;c__Cyanobacteri...</td>\n",
       "      <td>Malaysia: Sungkai; Perak</td>\n",
       "      <td>alkaline hot spring water</td>\n",
       "      <td>d__Bacteria;p__Cyanobacteriota;c__Cyanophyceae...</td>\n",
       "      <td>d__Bacteria;x__Terrabacteria group;x__Cyanobac...</td>\n",
       "      <td>MIITQGKINELLSEPGCEHNHHTHGDKKNKACKQQAQPGAAQGGCA...</td>\n",
       "      <td>GCVAYFRTHLSRHYKEPFSAVSSSMTEDAAVFGGLNNMIEGLQVSY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>GB_GCA_029266015.1</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>China:Yunnan province</td>\n",
       "      <td>native outdoor hot spring sediment</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__;o__;f__;g__;s__</td>\n",
       "      <td>d__Bacteria;x__FCB group;x__Bacteroidota/Chlor...</td>\n",
       "      <td>MKTKVTMDPERWKAEVLTKYPSKVAKKRAKSIVVNEGDEVKQIQAN...</td>\n",
       "      <td>MLLRHTTDKIIEREALTVNPAKTCQPVGAMYAALGIHGCLPHSHGS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1984</th>\n",
       "      <td>GB_GCA_030389965.1</td>\n",
       "      <td>d__Bacteria;p__Chloroflexota;c__Chloroflexia;o...</td>\n",
       "      <td>Canada: near Kenora; Ontario; Lake 227; IISD-E...</td>\n",
       "      <td>Single colony of Candidatus Chlorohelix alloph...</td>\n",
       "      <td>d__Bacteria;p__Chloroflexota;c__;o__;f__;g__;s__</td>\n",
       "      <td>d__Bacteria;x__Terrabacteria group;p__Chlorofl...</td>\n",
       "      <td>MAVLKCDATIPERDKHILIKGEGQDGKCSIACNSATTPGDMTERGC...</td>\n",
       "      <td>MSTVVKQERAVSINPLRACAPIGAMLATFGIHGALTINHGSQGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>GB_GCA_020049135.1</td>\n",
       "      <td>d__Bacteria;p__Verrucomicrobiota;c__Verrucomic...</td>\n",
       "      <td>USA: northwestern Lake Huron</td>\n",
       "      <td>purple microbial (photo- and chemosynthetic) m...</td>\n",
       "      <td>d__Bacteria;p__;c__;o__;f__;g__;s__</td>\n",
       "      <td>d__Bacteria;x__unclassified Bacteria;s__bacterium</td>\n",
       "      <td>MPTDIEMKNPDGIASSIPAEKIKEEMLKAYPAKTAKKRGKQILSND...</td>\n",
       "      <td>MNTVLEQEEEKEMKHCTATRNACKLCTPLGACLAFRGVEGCIPFLH...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1061 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GenomeID                                      gtdb_taxonomy  \\\n",
       "0     GB_GCA_903822005.1  d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "1     GB_GCA_937889015.1  d__Bacteria;p__Planctomycetota;c__Planctomycet...   \n",
       "2     GB_GCA_937924475.1  d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "3     GB_GCA_002339325.1  d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "4     GB_GCA_020725335.1  d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "...                  ...                                                ...   \n",
       "1978  GB_GCA_026418275.1  d__Bacteria;p__Bacillota_A;c__Clostridia;o__Ac...   \n",
       "1981  GB_GCA_026419705.1  d__Bacteria;p__Cyanobacteriota;c__Cyanobacteri...   \n",
       "1983  GB_GCA_029266015.1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "1984  GB_GCA_030389965.1  d__Bacteria;p__Chloroflexota;c__Chloroflexia;o...   \n",
       "1985  GB_GCA_020049135.1  d__Bacteria;p__Verrucomicrobiota;c__Verrucomic...   \n",
       "\n",
       "                                           ncbi_country  \\\n",
       "0                                                  none   \n",
       "1                                                  none   \n",
       "2                                                  none   \n",
       "3                             South Africa:Carltonville   \n",
       "4            Russia: Borehole near Beyelii Yar in Tomsk   \n",
       "...                                                 ...   \n",
       "1978                           Malaysia: Sungkai; Perak   \n",
       "1981                           Malaysia: Sungkai; Perak   \n",
       "1983                              China:Yunnan province   \n",
       "1984  Canada: near Kenora; Ontario; Lake 227; IISD-E...   \n",
       "1985                       USA: northwestern Lake Huron   \n",
       "\n",
       "                                  ncbi_isolation_source  \\\n",
       "0                                            lake water   \n",
       "1                                            metagenome   \n",
       "2                                      Mariane sediment   \n",
       "3                                                 water   \n",
       "4                groundwater filtered on 0.22 m filters   \n",
       "...                                                 ...   \n",
       "1978                          alkaline hot spring water   \n",
       "1981                          alkaline hot spring water   \n",
       "1983                 native outdoor hot spring sediment   \n",
       "1984  Single colony of Candidatus Chlorohelix alloph...   \n",
       "1985  purple microbial (photo- and chemosynthetic) m...   \n",
       "\n",
       "                                          ncbi_taxonomy  \\\n",
       "0     d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...   \n",
       "1     d__Bacteria;p__Planctomycetota;c__Planctomycet...   \n",
       "2                   d__Bacteria;p__;c__;o__;f__;g__;s__   \n",
       "3     d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...   \n",
       "4     d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "...                                                 ...   \n",
       "1978  d__Bacteria;p__Bacillota;c__Clostridia;o__;f__...   \n",
       "1981  d__Bacteria;p__Cyanobacteriota;c__Cyanophyceae...   \n",
       "1983    d__Bacteria;p__Bacteroidota;c__;o__;f__;g__;s__   \n",
       "1984   d__Bacteria;p__Chloroflexota;c__;o__;f__;g__;s__   \n",
       "1985                d__Bacteria;p__;c__;o__;f__;g__;s__   \n",
       "\n",
       "                               ncbi_taxonomy_unfiltered  \\\n",
       "0     d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...   \n",
       "1     d__Bacteria;x__PVC group;p__Planctomycetota;c_...   \n",
       "2     d__Bacteria;x__environmental samples;s__uncult...   \n",
       "3     d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...   \n",
       "4     d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "...                                                 ...   \n",
       "1978  d__Bacteria;x__Terrabacteria group;p__Bacillot...   \n",
       "1981  d__Bacteria;x__Terrabacteria group;x__Cyanobac...   \n",
       "1983  d__Bacteria;x__FCB group;x__Bacteroidota/Chlor...   \n",
       "1984  d__Bacteria;x__Terrabacteria group;p__Chlorofl...   \n",
       "1985  d__Bacteria;x__unclassified Bacteria;s__bacterium   \n",
       "\n",
       "                                               Seq_nifD  \\\n",
       "0     MSTAIDDTRKIIDEVLEAYPEKVRKARARHLAPNDPAGSCASCKIQ...   \n",
       "1     MAPELIDPERVKQELLQAYPAKVGRKRAGQMVVNRKDTPGESPEIV...   \n",
       "2     MNTYTINKKNEEEIRHHRVCRSRGGESCAFDGAMIVLQPIADAVHL...   \n",
       "3     MTRIIEQDTQRIVDEVLEVYPEKTKKERKKHLLANDPECSSSTCGL...   \n",
       "4     MPRIIEQDTEKLIEEVLEIYPEKTRKERRKHLISNDPECTTSTCQL...   \n",
       "...                                                 ...   \n",
       "1978  MPINIQNIEAATRENRLGAITGYHGTIKDLVSKATCCSLKNRERCF...   \n",
       "1981  MIITQGKINELLSEPGCEHNHHTHGDKKNKACKQQAQPGAAQGGCA...   \n",
       "1983  MKTKVTMDPERWKAEVLTKYPSKVAKKRAKSIVVNEGDEVKQIQAN...   \n",
       "1984  MAVLKCDATIPERDKHILIKGEGQDGKCSIACNSATTPGDMTERGC...   \n",
       "1985  MPTDIEMKNPDGIASSIPAEKIKEEMLKAYPAKTAKKRGKQILSND...   \n",
       "\n",
       "                                               Seq_nifK  \n",
       "0     MSKLKIKDHSELFLEPEYVDLFERKKEFEDACPDEEVKRVLDWTKT...  \n",
       "1     MNLLKGAVRGSQENPTVLLRHTTPNVVERQALTINPAKTCQPIGAM...  \n",
       "2     MFAISTTCMAEVIGDDLNAFIKKAKQQGVIPEQLPTPYSNTSSFTG...  \n",
       "3     MKVQDHVKLYRREDYEKLFRDKREFECPYPEEEVKKVEEYTKTEEY...  \n",
       "4     MKIMDHVELYKSEDYKELFENKKQFECPYSEEEVKKIEEYTKTEEY...  \n",
       "...                                                 ...  \n",
       "1978  MTQLIEQLRHVCSLGALQSVLAIHRAVPILHAGPGCGQKLWSATGV...  \n",
       "1981  GCVAYFRTHLSRHYKEPFSAVSSSMTEDAAVFGGLNNMIEGLQVSY...  \n",
       "1983  MLLRHTTDKIIEREALTVNPAKTCQPVGAMYAALGIHGCLPHSHGS...  \n",
       "1984  MSTVVKQERAVSINPLRACAPIGAMLATFGIHGALTINHGSQGCAT...  \n",
       "1985  MNTVLEQEEEKEMKHCTATRNACKLCTPLGACLAFRGVEGCIPFLH...  \n",
       "\n",
       "[1061 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get thermophiles (for Mikhail)\n",
    "\n",
    "# make sure seqs are in pruned tree (i.e. from nitrogenase clades)\n",
    "tree = Phylo.read(\"Nitrogenase_clades.txt\", \"newick\")\n",
    "\n",
    "# Extract and print node names\n",
    "nodes = []\n",
    "for clade in tree.get_terminals():\n",
    "    nodes.append(clade.name)\n",
    "\n",
    "# uncluster seqs (using tsv files)\n",
    "clusters = pd.read_csv('clustered_arsH_hmm_90_cluster.tsv', sep='\\t', header=None, index_col = 0)\n",
    "diazotrophs = clusters[clusters.index.isin(nodes)][1].to_list()\n",
    "\n",
    "# load GTDB metadata\n",
    "GTDB_metadata = pd.read_csv('GTDB_metadata.tsv', sep = '\\t', usecols=['accession', 'gtdb_taxonomy', 'ncbi_taxonomy', \n",
    "                                                'ncbi_taxonomy_unfiltered', 'ncbi_country', 'ncbi_isolation_source'])\n",
    "\n",
    "# get arsD seqs\n",
    "arsD_seqs = list(SeqIO.parse(\"filtered_arsD_LAMG.fasta\", \"fasta\"))\n",
    "arsD_df = pd.DataFrame([(record.id, str(record.seq)) for record in arsD_seqs], columns = ['GenomeID', 'Seq'])\n",
    "\n",
    "# get arsK seqs\n",
    "arsK_seqs = list(SeqIO.parse(\"filtered_arsK_LAMG.fasta\", \"fasta\"))\n",
    "arsK_df = pd.DataFrame([(record.id, str(record.seq)) for record in arsK_seqs], columns = ['GenomeID', 'Seq'])\n",
    "\n",
    "# merge dataframes to include accession, metadata, and sequences (for filtering)\n",
    "diazotrophs = pd.DataFrame(diazotrophs, columns = ['GenomeID'])\n",
    "diazotrophs = pd.merge(diazotrophs, GTDB_metadata, left_on = 'GenomeID', right_on = 'accession', how = 'left').drop(columns = 'accession')\n",
    "diazotrophs = pd.merge(diazotrophs, arsD_df, on = 'GenomeID', how = 'left')\n",
    "diazotrophs = pd.merge(diazotrophs, arsK_df, on = 'GenomeID', how = 'left', suffixes=('_arsD', '_arsK'))\n",
    "\n",
    "# keep species w/ \"therm\" or \"heat\" in taxonomy or isolation source\n",
    "thermophiles = diazotrophs[diazotrophs.gtdb_taxonomy.str.contains('therm', case = False)]\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_taxonomy.str.contains('therm', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_taxonomy_unfiltered.str.contains('therm', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_country.str.contains('therm', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_country.str.contains('hot', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_isolation_source.str.contains('therm', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_isolation_source.str.contains('hot', case = False)]], ignore_index=True)\n",
    "\n",
    "# remove duplicates\n",
    "thermophiles = thermophiles.drop_duplicates(subset = 'GenomeID')\n",
    "\n",
    "# export to csv\n",
    "thermophiles.to_csv('thermophiles_LAMG.csv', index = False, sep='\\t')\n",
    "\n",
    "thermophiles\n",
    "\n",
    "# awk '{print \">\" $1 \"|\" $2; print $NF}' thermophiles_LAMG.csv > thermo_arsK.fasta\n",
    "# awk '{print \">\" $1 \"|\" $2; print $(NF-1)}' thermophiles_LAMG.csv > thermo_arsD.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# information scoring\n",
    "\n",
    "# open aln file\n",
    "file = f\"../../thermo_arsD.aln\"\n",
    "\n",
    "# store in dataframe        \n",
    "results = list(SeqIO.parse(file, \"fasta\"))\n",
    "acc = [result.id.split('|')[0] for result in results]\n",
    "seqs = [list(str(result.seq)) for result in results]\n",
    "des = [[result.description.split('|')[-1]] for result in results]\n",
    "\n",
    "pssm = pd.DataFrame(seqs, index = acc)\n",
    "\n",
    "# get counts, frequency, entropy --> score\n",
    "score = pssm.apply(pd.Series.value_counts, axis = 0) # count\n",
    "score = score.apply(lambda x: x/1062, axis = 0) # frequency\n",
    "score.fillna(0, inplace = True) # replace Nan w/ 0\n",
    "score.loc['information'] = np.log2(22) + score.map(lambda x: x * np.log2(x) if x > 0 else 0).sum() # entropy, 0*log0 = 0\n",
    "score.iloc[:-1, :] = score.apply(lambda x: x*x.loc['information'], axis = 0).iloc[:-1,:] # individual score (closer to 4.39 = more conserved)\n",
    "\n",
    "# find sites of interest\n",
    "score.loc['azo'] = seqs[0]\n",
    "score.loc['max'] = score.iloc[1:-2, :].idxmax()\n",
    "sites = score.where(score.loc['max'] == score.loc['azo'], inplace=False)\n",
    "score.loc['azo-match'] = score.apply(lambda x: x.loc['max'] == x.loc['azo'])\n",
    "\n",
    "# sites of interest\n",
    "# sites = score[~score.loc['azo-match']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>1837</th>\n",
       "      <th>1838</th>\n",
       "      <th>1839</th>\n",
       "      <th>1840</th>\n",
       "      <th>1841</th>\n",
       "      <th>1842</th>\n",
       "      <th>1843</th>\n",
       "      <th>1844</th>\n",
       "      <th>1845</th>\n",
       "      <th>1846</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>...</td>\n",
       "      <td>4.330477</td>\n",
       "      <td>4.336893</td>\n",
       "      <td>4.332668</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028815</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004119</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004119</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004123</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008233</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.037103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004116</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.032948</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>information</th>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>...</td>\n",
       "      <td>4.371642</td>\n",
       "      <td>4.378118</td>\n",
       "      <td>4.373853</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azo</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>M</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>F</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>H</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>R</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>H</td>\n",
       "      <td>D</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azo-match</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26 rows × 1847 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0         1         2         3         4         5     \\\n",
       "-            4.444419  4.444419  4.444419  4.444419  4.444419  4.444419   \n",
       "A                 0.0       0.0       0.0       0.0       0.0  0.004189   \n",
       "C                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "D                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "E                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "F                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "G                 0.0       0.0       0.0       0.0  0.004189       0.0   \n",
       "H                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "I                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "K                 0.0  0.004189  0.004189       0.0       0.0       0.0   \n",
       "L                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "M            0.004189       0.0       0.0       0.0       0.0       0.0   \n",
       "N                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "P                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "Q                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "R                 0.0       0.0       0.0  0.004189       0.0       0.0   \n",
       "S                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "T                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "V                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "W                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "X                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "Y                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "information  4.448608  4.448608  4.448608  4.448608  4.448608  4.448608   \n",
       "azo                 -         -         -         -         -         -   \n",
       "max                 M         K         K         R         G         A   \n",
       "azo-match       False     False     False     False     False     False   \n",
       "\n",
       "                 6         7         8         9     ...      1837      1838  \\\n",
       "-            4.444419  4.444419  4.444419  4.444419  ...  4.330477  4.336893   \n",
       "A                 0.0       0.0       0.0       0.0  ...  0.028815       0.0   \n",
       "C                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "D                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "E                 0.0       0.0       0.0       0.0  ...       0.0  0.004123   \n",
       "F                 0.0  0.004189       0.0       0.0  ...       0.0       0.0   \n",
       "G                 0.0       0.0  0.004189       0.0  ...  0.008233       0.0   \n",
       "H                 0.0       0.0       0.0       0.0  ...       0.0  0.037103   \n",
       "I                 0.0       0.0       0.0       0.0  ...  0.004116       0.0   \n",
       "K                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "L                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "M                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "N                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "P                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "Q                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "R                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "S                 0.0       0.0       0.0  0.004189  ...       0.0       0.0   \n",
       "T                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "V            0.004189       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "W                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "X                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "Y                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "information  4.448608  4.448608  4.448608  4.448608  ...  4.371642  4.378118   \n",
       "azo                 -         -         -         -  ...         -         -   \n",
       "max                 V         F         G         S  ...         A         H   \n",
       "azo-match       False     False     False     False  ...     False     False   \n",
       "\n",
       "                 1839      1840      1841      1842      1843      1844  \\\n",
       "-            4.332668  4.444419  4.444419  4.444419  4.444419  4.444419   \n",
       "A                 0.0       0.0       0.0  0.004189       0.0       0.0   \n",
       "C            0.004119       0.0       0.0       0.0       0.0       0.0   \n",
       "D            0.004119       0.0       0.0       0.0       0.0       0.0   \n",
       "E                 0.0  0.004189       0.0       0.0       0.0       0.0   \n",
       "F                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "G                 0.0       0.0       0.0       0.0  0.004189       0.0   \n",
       "H                 0.0       0.0       0.0       0.0       0.0  0.004189   \n",
       "I                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "K                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "L                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "M                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "N            0.032948       0.0       0.0       0.0       0.0       0.0   \n",
       "P                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "Q                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "R                 0.0       0.0  0.004189       0.0       0.0       0.0   \n",
       "S                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "T                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "V                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "W                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "X                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "Y                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "information  4.373853  4.448608  4.448608  4.448608  4.448608  4.448608   \n",
       "azo                 -         -         -         -         -         -   \n",
       "max                 N         E         R         A         G         H   \n",
       "azo-match       False     False     False     False     False     False   \n",
       "\n",
       "                 1845      1846  \n",
       "-            4.444419  4.444419  \n",
       "A                 0.0       0.0  \n",
       "C                 0.0       0.0  \n",
       "D            0.004189       0.0  \n",
       "E                 0.0       0.0  \n",
       "F                 0.0       0.0  \n",
       "G                 0.0       0.0  \n",
       "H                 0.0       0.0  \n",
       "I                 0.0       0.0  \n",
       "K                 0.0       0.0  \n",
       "L                 0.0       0.0  \n",
       "M                 0.0       0.0  \n",
       "N                 0.0       0.0  \n",
       "P                 0.0       0.0  \n",
       "Q                 0.0       0.0  \n",
       "R                 0.0       0.0  \n",
       "S                 0.0       0.0  \n",
       "T                 0.0       0.0  \n",
       "V                 0.0  0.004189  \n",
       "W                 0.0       0.0  \n",
       "X                 0.0       0.0  \n",
       "Y                 0.0       0.0  \n",
       "information  4.448608  4.448608  \n",
       "azo                 -         -  \n",
       "max                 D         V  \n",
       "azo-match       False     False  \n",
       "\n",
       "[26 rows x 1847 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>172</th>\n",
       "      <th>174</th>\n",
       "      <th>308</th>\n",
       "      <th>475</th>\n",
       "      <th>485</th>\n",
       "      <th>488</th>\n",
       "      <th>495</th>\n",
       "      <th>502</th>\n",
       "      <th>527</th>\n",
       "      <th>569</th>\n",
       "      <th>...</th>\n",
       "      <th>1083</th>\n",
       "      <th>1126</th>\n",
       "      <th>1205</th>\n",
       "      <th>1220</th>\n",
       "      <th>1244</th>\n",
       "      <th>1271</th>\n",
       "      <th>1282</th>\n",
       "      <th>1286</th>\n",
       "      <th>1306</th>\n",
       "      <th>1314</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Conservation</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Quality</th>\n",
       "      <td>1135.2317</td>\n",
       "      <td>1278.8102</td>\n",
       "      <td>1664.9602</td>\n",
       "      <td>1139.5281</td>\n",
       "      <td>1677.4182</td>\n",
       "      <td>2033.7096</td>\n",
       "      <td>2126.2568</td>\n",
       "      <td>1651.0927</td>\n",
       "      <td>1581.4888</td>\n",
       "      <td>1248.7866</td>\n",
       "      <td>...</td>\n",
       "      <td>1086.1302</td>\n",
       "      <td>1352.8774</td>\n",
       "      <td>1797.6267</td>\n",
       "      <td>1080.2826</td>\n",
       "      <td>878.3229</td>\n",
       "      <td>205.22469</td>\n",
       "      <td>2041.5193</td>\n",
       "      <td>1996.0206</td>\n",
       "      <td>1603.496</td>\n",
       "      <td>1140.915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Consensus#</th>\n",
       "      <td>61.393597</td>\n",
       "      <td>54.613937</td>\n",
       "      <td>65.53672</td>\n",
       "      <td>50.282486</td>\n",
       "      <td>54.613937</td>\n",
       "      <td>58.568737</td>\n",
       "      <td>56.026367</td>\n",
       "      <td>90.20715</td>\n",
       "      <td>71.56309</td>\n",
       "      <td>56.40301</td>\n",
       "      <td>...</td>\n",
       "      <td>50.47081</td>\n",
       "      <td>67.3258</td>\n",
       "      <td>79.472694</td>\n",
       "      <td>56.967983</td>\n",
       "      <td>55.461395</td>\n",
       "      <td>53.483994</td>\n",
       "      <td>59.60452</td>\n",
       "      <td>53.107346</td>\n",
       "      <td>63.276836</td>\n",
       "      <td>52.259888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Consensus</th>\n",
       "      <td>K 61.39%; N 8.66%; D 8.19%; E 2.64%; R 1.69%;...</td>\n",
       "      <td>A 54.61%; K 11.49%; S 8.19%; E 4.90%; R 3.30%...</td>\n",
       "      <td>T 65.54%; S 18.46%; A 3.48%; V 3.48%; R 0.75%...</td>\n",
       "      <td>C 50.28%; A 31.54%; S 12.71%; M 0.47%; V 0.47...</td>\n",
       "      <td>R 54.61%; K 28.06%; S 4.14%; A 3.11%; G 2.35%...</td>\n",
       "      <td>V 58.57%; I 27.50%; L 7.91%; A 2.73%; F 0.47%...</td>\n",
       "      <td>I 56.03%; V 39.64%; A 1.41%; R 0.47%; S 0.47%...</td>\n",
       "      <td>W 90.21%; R 4.24%; S 0.94%; H 0.66%; Y 0.56%;...</td>\n",
       "      <td>T 71.56%; G 8.47%; S 8.10%; A 3.20%; Y 2.35%;...</td>\n",
       "      <td>N 56.40%; T 13.28%; S 12.62%; Q 7.82%; V 1.22...</td>\n",
       "      <td>...</td>\n",
       "      <td>A 50.47%; K 1.51%; S 1.51%; E 0.94%; R 0.85%;...</td>\n",
       "      <td>G 67.33%; A 14.60%; I 4.90%; D 3.77%; S 1.51%...</td>\n",
       "      <td>Y 79.47%; F 14.78%; P 1.98%; V 0.38%; L 0.28%...</td>\n",
       "      <td>Y 56.97%; W 35.40%; H 2.26%; Q 1.41%; D 0.94%...</td>\n",
       "      <td>A 55.46%; H 33.99%; I 1.69%; V 1.51%; T 1.32%...</td>\n",
       "      <td>N 53.48%; I 21.19%; V 13.84%; T 2.17%; R 1.69...</td>\n",
       "      <td>I 59.60%; M 19.96%; V 12.05%; L 6.97%; W 0.19...</td>\n",
       "      <td>V 53.11%; I 22.13%; L 17.23%; M 1.79%; F 1.69...</td>\n",
       "      <td>S 63.28%; N 10.08%; A 9.32%; T 8.38%; C 1.41%...</td>\n",
       "      <td>K 52.26%; R 11.02%; T 7.34%; E 3.95%; Q 3.30%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Occupancy</th>\n",
       "      <td>938</td>\n",
       "      <td>1018</td>\n",
       "      <td>984</td>\n",
       "      <td>1027</td>\n",
       "      <td>1040</td>\n",
       "      <td>1048</td>\n",
       "      <td>1054</td>\n",
       "      <td>1050</td>\n",
       "      <td>1042</td>\n",
       "      <td>1033</td>\n",
       "      <td>...</td>\n",
       "      <td>610</td>\n",
       "      <td>1040</td>\n",
       "      <td>1045</td>\n",
       "      <td>1054</td>\n",
       "      <td>1050</td>\n",
       "      <td>1045</td>\n",
       "      <td>1052</td>\n",
       "      <td>1053</td>\n",
       "      <td>1034</td>\n",
       "      <td>962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>K</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>R</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>S</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azo</th>\n",
       "      <td>D</td>\n",
       "      <td>N</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>K</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>S</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>N</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azo_match</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           172   \\\n",
       "0                                                                 \n",
       "Conservation                                                  1   \n",
       "Quality                                               1135.2317   \n",
       "Consensus#                                            61.393597   \n",
       "Consensus      K 61.39%; N 8.66%; D 8.19%; E 2.64%; R 1.69%;...   \n",
       "Occupancy                                                   938   \n",
       "max                                                           K   \n",
       "azo                                                           D   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           174   \\\n",
       "0                                                                 \n",
       "Conservation                                                  0   \n",
       "Quality                                               1278.8102   \n",
       "Consensus#                                            54.613937   \n",
       "Consensus      A 54.61%; K 11.49%; S 8.19%; E 4.90%; R 3.30%...   \n",
       "Occupancy                                                  1018   \n",
       "max                                                           A   \n",
       "azo                                                           N   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           308   \\\n",
       "0                                                                 \n",
       "Conservation                                                  1   \n",
       "Quality                                               1664.9602   \n",
       "Consensus#                                             65.53672   \n",
       "Consensus      T 65.54%; S 18.46%; A 3.48%; V 3.48%; R 0.75%...   \n",
       "Occupancy                                                   984   \n",
       "max                                                           T   \n",
       "azo                                                           S   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           475   \\\n",
       "0                                                                 \n",
       "Conservation                                                  1   \n",
       "Quality                                               1139.5281   \n",
       "Consensus#                                            50.282486   \n",
       "Consensus      C 50.28%; A 31.54%; S 12.71%; M 0.47%; V 0.47...   \n",
       "Occupancy                                                  1027   \n",
       "max                                                           C   \n",
       "azo                                                           S   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           485   \\\n",
       "0                                                                 \n",
       "Conservation                                                  4   \n",
       "Quality                                               1677.4182   \n",
       "Consensus#                                            54.613937   \n",
       "Consensus      R 54.61%; K 28.06%; S 4.14%; A 3.11%; G 2.35%...   \n",
       "Occupancy                                                  1040   \n",
       "max                                                           R   \n",
       "azo                                                           K   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           488   \\\n",
       "0                                                                 \n",
       "Conservation                                                  9   \n",
       "Quality                                               2033.7096   \n",
       "Consensus#                                            58.568737   \n",
       "Consensus      V 58.57%; I 27.50%; L 7.91%; A 2.73%; F 0.47%...   \n",
       "Occupancy                                                  1048   \n",
       "max                                                           V   \n",
       "azo                                                           I   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           495   \\\n",
       "0                                                                 \n",
       "Conservation                                                  9   \n",
       "Quality                                               2126.2568   \n",
       "Consensus#                                            56.026367   \n",
       "Consensus      I 56.03%; V 39.64%; A 1.41%; R 0.47%; S 0.47%...   \n",
       "Occupancy                                                  1054   \n",
       "max                                                           I   \n",
       "azo                                                           V   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           502   \\\n",
       "0                                                                 \n",
       "Conservation                                                  6   \n",
       "Quality                                               1651.0927   \n",
       "Consensus#                                             90.20715   \n",
       "Consensus      W 90.21%; R 4.24%; S 0.94%; H 0.66%; Y 0.56%;...   \n",
       "Occupancy                                                  1050   \n",
       "max                                                           W   \n",
       "azo                                                           R   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           527   \\\n",
       "0                                                                 \n",
       "Conservation                                                  7   \n",
       "Quality                                               1581.4888   \n",
       "Consensus#                                             71.56309   \n",
       "Consensus      T 71.56%; G 8.47%; S 8.10%; A 3.20%; Y 2.35%;...   \n",
       "Occupancy                                                  1042   \n",
       "max                                                           T   \n",
       "azo                                                           G   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           569   ...  \\\n",
       "0                                                                ...   \n",
       "Conservation                                                  7  ...   \n",
       "Quality                                               1248.7866  ...   \n",
       "Consensus#                                             56.40301  ...   \n",
       "Consensus      N 56.40%; T 13.28%; S 12.62%; Q 7.82%; V 1.22...  ...   \n",
       "Occupancy                                                  1033  ...   \n",
       "max                                                           N  ...   \n",
       "azo                                                           A  ...   \n",
       "azo_match                                                  True  ...   \n",
       "\n",
       "                                                           1083  \\\n",
       "0                                                                 \n",
       "Conservation                                                  0   \n",
       "Quality                                               1086.1302   \n",
       "Consensus#                                             50.47081   \n",
       "Consensus      A 50.47%; K 1.51%; S 1.51%; E 0.94%; R 0.85%;...   \n",
       "Occupancy                                                   610   \n",
       "max                                                           A   \n",
       "azo                                                           -   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1126  \\\n",
       "0                                                                 \n",
       "Conservation                                                  3   \n",
       "Quality                                               1352.8774   \n",
       "Consensus#                                              67.3258   \n",
       "Consensus      G 67.33%; A 14.60%; I 4.90%; D 3.77%; S 1.51%...   \n",
       "Occupancy                                                  1040   \n",
       "max                                                           G   \n",
       "azo                                                           S   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1205  \\\n",
       "0                                                                 \n",
       "Conservation                                                  9   \n",
       "Quality                                               1797.6267   \n",
       "Consensus#                                            79.472694   \n",
       "Consensus      Y 79.47%; F 14.78%; P 1.98%; V 0.38%; L 0.28%...   \n",
       "Occupancy                                                  1045   \n",
       "max                                                           Y   \n",
       "azo                                                           F   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1220  \\\n",
       "0                                                                 \n",
       "Conservation                                                 10   \n",
       "Quality                                               1080.2826   \n",
       "Consensus#                                            56.967983   \n",
       "Consensus      Y 56.97%; W 35.40%; H 2.26%; Q 1.41%; D 0.94%...   \n",
       "Occupancy                                                  1054   \n",
       "max                                                           Y   \n",
       "azo                                                           W   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1244  \\\n",
       "0                                                                 \n",
       "Conservation                                                  4   \n",
       "Quality                                                878.3229   \n",
       "Consensus#                                            55.461395   \n",
       "Consensus      A 55.46%; H 33.99%; I 1.69%; V 1.51%; T 1.32%...   \n",
       "Occupancy                                                  1050   \n",
       "max                                                           A   \n",
       "azo                                                           H   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1271  \\\n",
       "0                                                                 \n",
       "Conservation                                                  6   \n",
       "Quality                                               205.22469   \n",
       "Consensus#                                            53.483994   \n",
       "Consensus      N 53.48%; I 21.19%; V 13.84%; T 2.17%; R 1.69...   \n",
       "Occupancy                                                  1045   \n",
       "max                                                           N   \n",
       "azo                                                           I   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1282  \\\n",
       "0                                                                 \n",
       "Conservation                                                  8   \n",
       "Quality                                               2041.5193   \n",
       "Consensus#                                             59.60452   \n",
       "Consensus      I 59.60%; M 19.96%; V 12.05%; L 6.97%; W 0.19...   \n",
       "Occupancy                                                  1052   \n",
       "max                                                           I   \n",
       "azo                                                           M   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1286  \\\n",
       "0                                                                 \n",
       "Conservation                                                  9   \n",
       "Quality                                               1996.0206   \n",
       "Consensus#                                            53.107346   \n",
       "Consensus      V 53.11%; I 22.13%; L 17.23%; M 1.79%; F 1.69...   \n",
       "Occupancy                                                  1053   \n",
       "max                                                           V   \n",
       "azo                                                           L   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1306  \\\n",
       "0                                                                 \n",
       "Conservation                                                  7   \n",
       "Quality                                                1603.496   \n",
       "Consensus#                                            63.276836   \n",
       "Consensus      S 63.28%; N 10.08%; A 9.32%; T 8.38%; C 1.41%...   \n",
       "Occupancy                                                  1034   \n",
       "max                                                           S   \n",
       "azo                                                           N   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1314  \n",
       "0                                                                \n",
       "Conservation                                                  1  \n",
       "Quality                                                1140.915  \n",
       "Consensus#                                            52.259888  \n",
       "Consensus      K 52.26%; R 11.02%; T 7.34%; E 3.95%; Q 3.30%...  \n",
       "Occupancy                                                   962  \n",
       "max                                                           K  \n",
       "azo                                                           Q  \n",
       "azo_match                                                  True  \n",
       "\n",
       "[8 rows x 59 columns]"
      ]
     },
     "execution_count": 393,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = f\"thermo_arsD_features.csv\"\n",
    "\n",
    "features = pd.read_csv(file, header=None, index_col = 0)\n",
    "features.columns = range(len(features.columns))\n",
    "features = features.drop(features.columns[-1], axis=1)\n",
    "features.loc['max'] = features.loc['Consensus']\n",
    "features.loc['Quality'] = features.apply(lambda x: float(x.loc['Quality']), axis=0)\n",
    "features.loc['Consensus#'] = features.apply(lambda x: float(x.loc['Consensus#']), axis=0)\n",
    "features.loc['max'] = features.apply(lambda x: x.loc['Consensus'].strip()[0], axis=0)\n",
    "features.loc['azo'] = seqs[0]\n",
    "features.loc['azo_match'] = score.apply(lambda x: x.loc['max'] != x.loc['azo'])\n",
    "features = features.loc[:, features.loc['azo_match']]\n",
    "\n",
    "features = features.loc[:, features.loc['Consensus#'] >= 50]\n",
    "# features = features.loc[:, features.loc['azo'] = 'L']\n",
    "# features = features.loc[:, features.loc['azo'] == 'I']\n",
    "# # features = features.loc[:, features.loc['azo'] != 'V']\n",
    "features#.to_csv('thermo_arsD_features_filt.csv', index = False, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>502</th>\n",
       "      <th>527</th>\n",
       "      <th>569</th>\n",
       "      <th>691</th>\n",
       "      <th>836</th>\n",
       "      <th>877</th>\n",
       "      <th>997</th>\n",
       "      <th>1314</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RS_GCF_001687335.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>Y</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_002898515.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_002093115.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_011207455.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>Y</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_027925145.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>L</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_903896745.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000195755.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000266945.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_030065275.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>L</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_009688965.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_947259715.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000711225.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_002423105.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>Q</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_025355965.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>K</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_030154185.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>V</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_015231595.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>L</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_003232385.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_023512345.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_030913705.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_003249225.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_003245715.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>Q</td>\n",
       "      <td>G</td>\n",
       "      <td>L</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   502  527  569  691  836  877  997  1314\n",
       "RS_GCF_001687335.1    W    T    Y    G    V    A    V    K\n",
       "RS_GCF_002898515.1    W    T    S    G    V    C    V    K\n",
       "RS_GCF_002093115.1    W    T    S    G    I    C    V    K\n",
       "RS_GCF_011207455.1    W    T    Y    G    V    A    V    R\n",
       "RS_GCF_027925145.1    W    T    N    G    L    A    V    K\n",
       "GB_GCA_903896745.1    W    T    N    G    I    A    V    K\n",
       "RS_GCF_000195755.1    W    T    N    G    V    A    V    R\n",
       "RS_GCF_000266945.1    W    T    N    G    V    A    V    K\n",
       "GB_GCA_030065275.1    W    T    N    G    L    A    V    R\n",
       "RS_GCF_009688965.1    W    T    N    G    I    A    V    R\n",
       "GB_GCA_947259715.1    W    T    -    G    V    C    V    K\n",
       "RS_GCF_000711225.1    W    T    C    G    V    C    V    K\n",
       "GB_GCA_002423105.1    W    T    Q    G    V    C    V    K\n",
       "GB_GCA_025355965.1    W    T    K    G    V    C    V    K\n",
       "GB_GCA_030154185.1    W    T    V    G    V    A    V    K\n",
       "GB_GCA_015231595.1    W    T    N    G    L    G    V    K\n",
       "GB_GCA_003232385.1    W    T    S    G    V    A    V    K\n",
       "GB_GCA_023512345.1    W    T    N    G    V    V    V    K\n",
       "RS_GCF_030913705.1    W    T    C    G    V    A    V    K\n",
       "GB_GCA_003249225.1    W    T    S    G    V    A    V    R\n",
       "GB_GCA_003245715.1    W    T    Q    G    L    A    V    K"
      ]
     },
     "execution_count": 436,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# thermo to plot MSA\n",
    "\n",
    "file = f\"../Thermo/thermo_arsD.aln\"\n",
    "\n",
    "# store in dataframe        \n",
    "results = list(SeqIO.parse(file, \"fasta\"))\n",
    "acc = [result.id.split('|')[0] for result in results]\n",
    "seqs = [list(str(result.seq)) for result in results]\n",
    "des = [[result.description.split('|')[-1]] for result in results]\n",
    "\n",
    "pssm = pd.DataFrame(seqs, index = acc)\n",
    "pssm = pssm[pssm.loc[:,502] !='R']\n",
    "pssm = pssm[pssm.loc[:,527] =='T']\n",
    "pssm = pssm[(pssm.loc[:,569] !='A')]# | (pssm.loc[:,569] =='T')]\n",
    "pssm = pssm[pssm.loc[:,691] =='G']\n",
    "pssm = pssm[(pssm.loc[:,836] =='V') | (pssm.loc[:,836] =='L')| (pssm.loc[:,836] =='I')]\n",
    "pssm = pssm[(pssm.loc[:,877] !='S') & (pssm.loc[:,877] !='T')]\n",
    "pssm = pssm[pssm.loc[:,997] !='I']\n",
    "pssm = pssm[(pssm.loc[:,1314] =='K') | (pssm.loc[:,1314] =='R')]\n",
    "pssm = pssm[[502, 527, 569, 691, 836, 877, 997, 1314]]\n",
    "pssm = pssm.drop_duplicates()\n",
    "pssm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21 16 39 36 33\n"
     ]
    }
   ],
   "source": [
    "file = f\"../Thermo/thermo_arsK.aln\"\n",
    "\n",
    "# store in dataframe        \n",
    "results = list(SeqIO.parse(file, \"fasta\"))\n",
    "seqs = [list(str(result.seq)) for result in results]\n",
    "protein = seqs[0]\n",
    "R = protein.count('R')\n",
    "H = protein.count('H')\n",
    "K = protein.count('K')\n",
    "D = protein.count('D')\n",
    "E = protein.count('E')\n",
    "\n",
    "print(R,H,K,D,E)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5592\n",
      "5592\n"
     ]
    }
   ],
   "source": [
    "file = '../operon-org/structure.pickle'\n",
    "import pickle\n",
    "\n",
    "file = pickle.load(open(file, 'rb'))\n",
    "print(len(file))\n",
    "\n",
    "remove_list = ['GB_GCA_002898135.1_BDTQ01000234.1', \n",
    "             'GB_GCA_021797365.1_JAJZZU010000300.1',\n",
    "             'GB_GCA_009649835.1_PYCK01000054.1',\n",
    "             'GB_GCA_014380135.1_JACMKN010000012.1',\n",
    "             'GB_GCA_017889725.1_JAGGAE010000092.1',\n",
    "             'GB_GCA_902385025.1_CABMDX010000290.1',\n",
    "             'GB_GCA_027004315.1_JALTHV010000232.1',\n",
    "             'GB_GCA_031288575.1_JAITRD010000305.1',\n",
    "             'GB_GCA_937141995.1_CAKZQX010000080.1',\n",
    "             'GB_GCA_029778665.1_JAKOPX010000087.1',\n",
    "             'GB_GCA_031276475.1_JAISMZ010000161.1',\n",
    "             'RS_GCF_015731765.1_NZ_VRYY01000094.1',\n",
    "             'GB_GCA_013041765.1_JABDRL010000192.1',\n",
    "             'GB_GCA_021798485.1_JAJZGK010000148.1',\n",
    "             'GB_GCA_031290695.1_JAITNB010000026.1',\n",
    "             'GB_GCA_031278575.1_JAITKD010000016.1',\n",
    "             'GB_GCA_937886325.1_CALABV010000158.1',\n",
    "             'GB_GCA_937915285.1_CALDNG010000494.1',\n",
    "             'GB_GCA_021802595.1_JAJZRT010000330.1']\n",
    "\n",
    "for item in remove_list:\n",
    "    to_remove = '../operon-org/input-fastas/' + item + '_operon.fasta'\n",
    "    try:\n",
    "        file.pop(to_remove)\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "pickle.dump(file, open('../operon-org/structure-update.pickle', 'wb'))\n",
    "print(len(file))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pro-pred",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
